// mat-range-datepicker/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export { NativeDateModule, MatRangeNativeDateModule, MAT_DATE_LOCALE, MAT_DATE_LOCALE_FACTORY, MAT_DATE_LOCALE_PROVIDER, DateAdapter, MAT_DATE_FORMATS, NativeDateAdapter, MAT_NATIVE_DATE_FORMATS } from './datetime/index';
export { MatRangeDatepickerModule, SatCalendar, SatCalendarHeader, SatCalendarCell, SatCalendarBody, MAT_DATEPICKER_SCROLL_STRATEGY, MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY, matRangeDatepickerContentBase, _matRangeDatepickerContentMixinBase, matRangeDatepickerContent, matRangeDatepicker, matDatepickerAnimations, MAT_DATEPICKER_VALUE_ACCESSOR, MAT_DATEPICKER_VALIDATORS, matRangeDatepickerInputEvent, matRangeDatepickerInput, matRangeDatepickerIntl, matRangeDatepickerToggleIcon, matRangeDatepickerToggle, SatMonthView, SatYearView } from './datepicker/index';
