// mat-range-datepicker/esm2015/datepicker/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export { MatRangeDatepickerModule } from './datepicker-module';
export { SatCalendar, SatCalendarHeader } from './calendar';
export { SatCalendarCell, SatCalendarBody } from './calendar-body';
export { MAT_DATEPICKER_SCROLL_STRATEGY, MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY, matRangeDatepickerContentBase, _matRangeDatepickerContentMixinBase, matRangeDatepickerContent, matRangeDatepicker } from './datepicker';
export { matDatepickerAnimations } from './datepicker-animations';
export { MAT_DATEPICKER_VALUE_ACCESSOR, MAT_DATEPICKER_VALIDATORS, matRangeDatepickerInputEvent, matRangeDatepickerInput } from './datepicker-input';
export { matRangeDatepickerIntl } from './datepicker-intl';
export { matRangeDatepickerToggleIcon, matRangeDatepickerToggle } from './datepicker-toggle';
export { SatMonthView } from './month-view';
export { SatYearView } from './year-view';
