// mat-range-datepicker/bundles/mat-range-datepicker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('tslib'), require('@angular/cdk/platform'), require('rxjs/operators'), require('@angular/cdk/keycodes'), require('@angular/cdk/bidi'), require('@angular/cdk/portal'), require('@angular/animations'), require('@angular/cdk/coercion'), require('@angular/cdk/overlay'), require('@angular/common'), require('@angular/material/core'), require('@angular/material/dialog'), require('@angular/forms'), require('@angular/material/form-field'), require('@angular/material/input'), require('@angular/cdk/a11y'), require('@angular/material/button')) :
    typeof define === 'function' && define.amd ? define('mat-range-datepicker', ['exports', '@angular/core', 'rxjs', 'tslib', '@angular/cdk/platform', 'rxjs/operators', '@angular/cdk/keycodes', '@angular/cdk/bidi', '@angular/cdk/portal', '@angular/animations', '@angular/cdk/coercion', '@angular/cdk/overlay', '@angular/common', '@angular/material/core', '@angular/material/dialog', '@angular/forms', '@angular/material/form-field', '@angular/material/input', '@angular/cdk/a11y', '@angular/material/button'], factory) :
    (factory((global['mat-range-datepicker'] = {}),global.ng.core,global.rxjs,global.tslib,global.ng.cdk.platform,global.Rx.Observable.prototype,global.ng.cdk.keycodes,global.ng.cdk.bidi,global.ng.cdk.portal,global.ng.animations,global.ng.cdk.coercion,global.ng.cdk.overlay,global.ng.common,global.ng.material.core,global.ng.material.dialog,global.ng.forms,global.ng.material['form-field'],global.ng.material.input,global.ng.cdk.a11y,global.ng.material.button));
}(this, (function (exports,core,rxjs,tslib,platform,operators,keycodes,bidi,portal,animations,coercion,overlay,common,core$1,dialog,forms,formField,input,a11y,button) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * InjectionToken for datepicker that can be used to override default locale code.
     */
    var /** @type {?} */ MAT_DATE_LOCALE = new core.InjectionToken('MAT_DATE_LOCALE', {
        providedIn: 'root',
        factory: MAT_DATE_LOCALE_FACTORY,
    });
    /**
     * \@docs-private
     * @return {?}
     */
    function MAT_DATE_LOCALE_FACTORY() {
        return core.inject(core.LOCALE_ID);
    }
    /**
     * No longer needed since MAT_DATE_LOCALE has been changed to a scoped injectable.
     * If you are importing and providing this in your code you can simply remove it.
     * @deprecated
     * \@deletion-target 7.0.0
     */
    var /** @type {?} */ MAT_DATE_LOCALE_PROVIDER = { provide: MAT_DATE_LOCALE, useExisting: core.LOCALE_ID };
    /**
     * Adapts type `D` to be usable as a date by cdk-based components that work with dates.
     * @abstract
     * @template D
     */
    var DateAdapter = (function () {
        function DateAdapter() {
            this._localeChanges = new rxjs.Subject();
        }
        Object.defineProperty(DateAdapter.prototype, "localeChanges", {
            /**
             * A stream that emits when the locale changes.
             * @return {?}
             */
            get: function () { return this._localeChanges; },
            enumerable: true,
            configurable: true
        });
        /**
         * Attempts to deserialize a value to a valid date object. This is different from parsing in that
         * deserialize should only accept non-ambiguous, locale-independent formats (e.g. a ISO 8601
         * string). The default implementation does not allow any deserialization, it simply checks that
         * the given value is already a valid date object or null. The `<mat-range-datepicker>` will call this
         * method on all of it's `\@Input()` properties that accept dates. It is therefore possible to
         * support passing values from your backend directly to these properties by overriding this method
         * to also deserialize the format used by your backend.
         * @param {?} value The value to be deserialized into a date object.
         * @return {?} The deserialized date object, either a valid date, null if the value can be
         *     deserialized into a null date (e.g. the empty string), or an invalid date.
         */
        DateAdapter.prototype.deserialize = function (value) {
            if (value == null || this.isDateInstance(value) && this.isValid(value)) {
                return value;
            }
            return this.invalid();
        };
        /**
         * Sets the locale used for all dates.
         * @param {?} locale The new locale.
         * @return {?}
         */
        DateAdapter.prototype.setLocale = function (locale) {
            this.locale = locale;
            this._localeChanges.next();
        };
        /**
         * Compares two dates.
         * @param {?} first The first date to compare.
         * @param {?} second The second date to compare.
         * @return {?} 0 if the dates are equal, a number less than 0 if the first date is earlier,
         *     a number greater than 0 if the first date is later.
         */
        DateAdapter.prototype.compareDate = function (first, second) {
            return this.getYear(first) - this.getYear(second) ||
                this.getMonth(first) - this.getMonth(second) ||
                this.getDate(first) - this.getDate(second);
        };
        /**
         * Checks if two dates are equal.
         * @param {?} first The first date to check.
         * @param {?} second The second date to check.
         * @return {?} Whether the two dates are equal.
         *     Null dates are considered equal to other null dates.
         */
        DateAdapter.prototype.sameDate = function (first, second) {
            if (first && second) {
                var /** @type {?} */ firstValid = this.isValid(first);
                var /** @type {?} */ secondValid = this.isValid(second);
                if (firstValid && secondValid) {
                    return !this.compareDate(first, second);
                }
                return firstValid == secondValid;
            }
            return first == second;
        };
        /**
         * Clamp the given date between min and max dates.
         * @param {?} date The date to clamp.
         * @param {?=} min The minimum value to allow. If null or omitted no min is enforced.
         * @param {?=} max The maximum value to allow. If null or omitted no max is enforced.
         * @return {?} `min` if `date` is less than `min`, `max` if date is greater than `max`,
         *     otherwise `date`.
         */
        DateAdapter.prototype.clampDate = function (date, min, max) {
            if (min && this.compareDate(date, min) < 0) {
                return min;
            }
            if (max && this.compareDate(date, max) > 0) {
                return max;
            }
            return date;
        };
        return DateAdapter;
    }());
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ MAT_DATE_FORMATS = new core.InjectionToken('mat-date-formats');
    /**
     * Whether the browser supports the Intl API.
     */
    var /** @type {?} */ SUPPORTS_INTL_API = typeof Intl != 'undefined';
    /**
     * The default month names to use if Intl API is not available.
     */
    var /** @type {?} */ DEFAULT_MONTH_NAMES = {
        'long': [
            'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September',
            'October', 'November', 'December'
        ],
        'short': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
        'narrow': ['J', 'F', 'M', 'A', 'M', 'J', 'J', 'A', 'S', 'O', 'N', 'D']
    };
    var ɵ0 = function (i) { return String(i + 1); };
    /**
     * The default date names to use if Intl API is not available.
     */
    var /** @type {?} */ DEFAULT_DATE_NAMES = range(31, ɵ0);
    /**
     * The default day of the week names to use if Intl API is not available.
     */
    var /** @type {?} */ DEFAULT_DAY_OF_WEEK_NAMES = {
        'long': ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
        'short': ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
        'narrow': ['S', 'M', 'T', 'W', 'T', 'F', 'S']
    };
    /**
     * First day of week according locale.
     * Taken form moment.js source code https://github.com/moment/moment/tree/develop/src/locale
     */
    var /** @type {?} */ FIRST_DAY_OF_WEEK = {
        af: 1, ar: 6, 'ar-ly': 6, 'ar-ma': 6, 'ar-tn': 1, az: 1, be: 1, bg: 1, bm: 1, br: 1, bs: 1, ca: 1, cs: 1, cv: 1,
        cy: 1, da: 1, de: 1, 'de-at': 1, 'de-ch': 1, el: 1, 'en-au': 1, 'en-gb': 1, 'en-ie': 1, 'en-nz': 1, eo: 1,
        es: 1, 'es-do': 1, et: 1, eu: 1, fa: 6, fi: 1, fo: 1, fr: 1, 'fr-ch': 1, fy: 1, gd: 1, gl: 1, 'gom-latn': 1,
        hr: 1, hu: 1, 'hy-am': 1, id: 1, is: 1, it: 1, jv: 1, ka: 1, kk: 1, km: 1, ky: 1, lb: 1, lt: 1, lv: 1, me: 1,
        mi: 1, mk: 1, ms: 1, 'ms-my': 1, mt: 1, my: 1, nb: 1, nl: 1, 'nl-be': 1, nn: 1, pl: 1, pt: 1, ro: 1, ru: 1,
        sd: 1, se: 1, sk: 1, sl: 1, sq: 1, sr: 1, 'sr-cyrl': 1, ss: 1, sv: 1, sw: 1, 'tet': 1, tg: 1, 'tl-ph': 1,
        'tlh': 1, tr: 1, 'tzl': 1, 'tzm': 6, 'tzm-latn': 6, 'ug-cn': 1, uk: 1, ur: 1, uz: 1, 'uz-latn': 1, vi: 1,
        'x-pseudo': 1, yo: 1, 'zh-cn': 1,
    };
    /**
     * Matches strings that have the form of a valid RFC 3339 string
     * (https://tools.ietf.org/html/rfc3339). Note that the string may not actually be a valid date
     * because the regex will match strings an with out of bounds month, date, etc.
     */
    var /** @type {?} */ ISO_8601_REGEX = /^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|(?:(?:\+|-)\d{2}:\d{2}))?)?$/;
    /**
     * Creates an array and fills it with values.
     * @template T
     * @param {?} length
     * @param {?} valueFunction
     * @return {?}
     */
    function range(length, valueFunction) {
        var /** @type {?} */ valuesArray = Array(length);
        for (var /** @type {?} */ i = 0; i < length; i++) {
            valuesArray[i] = valueFunction(i);
        }
        return valuesArray;
    }
    /**
     * Adapts the native JS Date for use with cdk-based components that work with dates.
     */
    var NativeDateAdapter = (function (_super) {
        tslib.__extends(NativeDateAdapter, _super);
        /**
         * @param {?} matDateLocale
         * @param {?} platform
         */
        function NativeDateAdapter(matDateLocale, platform$$1) {
            var _this = _super.call(this) || this;
            /**
             * Whether to use `timeZone: 'utc'` with `Intl.DateTimeFormat` when formatting dates.
             * Without this `Intl.DateTimeFormat` sometimes chooses the wrong timeZone, which can throw off
             * the result. (e.g. in the en-US locale `new Date(1800, 7, 14).toLocaleDateString()`
             * will produce `'8/13/1800'`.
             *
             * TODO(mmalerba): drop this variable. It's not being used in the code right now. We're now
             * getting the string representation of a Date object from it's utc representation. We're keeping
             * it here for sometime, just for precaution, in case we decide to revert some of these changes
             * though.
             */
            _this.useUtcForDisplay = true;
            _super.prototype.setLocale.call(_this, matDateLocale);
            // IE does its own time zone correction, so we disable this on IE.
            _this.useUtcForDisplay = !platform$$1.TRIDENT;
            _this._clampDate = platform$$1.TRIDENT || platform$$1.EDGE;
            return _this;
        }
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getYear = function (date) {
            return date.getFullYear();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getMonth = function (date) {
            return date.getMonth();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getDate = function (date) {
            return date.getDate();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getDayOfWeek = function (date) {
            return date.getDay();
        };
        /**
         * @param {?} style
         * @return {?}
         */
        NativeDateAdapter.prototype.getMonthNames = function (style$$1) {
            var _this = this;
            if (SUPPORTS_INTL_API) {
                var /** @type {?} */ dtf_1 = new Intl.DateTimeFormat(this.locale, { month: style$$1, timeZone: 'utc' });
                return range(12, function (i) { return _this._stripDirectionalityCharacters(_this._format(dtf_1, new Date(2017, i, 1))); });
            }
            return DEFAULT_MONTH_NAMES[style$$1];
        };
        /**
         * @return {?}
         */
        NativeDateAdapter.prototype.getDateNames = function () {
            var _this = this;
            if (SUPPORTS_INTL_API) {
                var /** @type {?} */ dtf_2 = new Intl.DateTimeFormat(this.locale, { day: 'numeric', timeZone: 'utc' });
                return range(31, function (i) { return _this._stripDirectionalityCharacters(_this._format(dtf_2, new Date(2017, 0, i + 1))); });
            }
            return DEFAULT_DATE_NAMES;
        };
        /**
         * @param {?} style
         * @return {?}
         */
        NativeDateAdapter.prototype.getDayOfWeekNames = function (style$$1) {
            var _this = this;
            if (SUPPORTS_INTL_API) {
                var /** @type {?} */ dtf_3 = new Intl.DateTimeFormat(this.locale, { weekday: style$$1, timeZone: 'utc' });
                return range(7, function (i) { return _this._stripDirectionalityCharacters(_this._format(dtf_3, new Date(2017, 0, i + 1))); });
            }
            return DEFAULT_DAY_OF_WEEK_NAMES[style$$1];
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getYearName = function (date) {
            if (SUPPORTS_INTL_API) {
                var /** @type {?} */ dtf = new Intl.DateTimeFormat(this.locale, { year: 'numeric', timeZone: 'utc' });
                return this._stripDirectionalityCharacters(this._format(dtf, date));
            }
            return String(this.getYear(date));
        };
        /**
         * @return {?}
         */
        NativeDateAdapter.prototype.getFirstDayOfWeek = function () {
            // We can't tell using native JS Date what the first day of the week is.
            // Sometimes people use excess language definition, e.g. ru-RU,
            // so we use fallback to two-letter language code
            var /** @type {?} */ locale = this.locale.toLowerCase();
            return FIRST_DAY_OF_WEEK[locale] || FIRST_DAY_OF_WEEK[locale.substr(0, 2)] || 0;
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.getNumDaysInMonth = function (date) {
            return this.getDate(this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + 1, 0));
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.clone = function (date) {
            return this.createDate(this.getYear(date), this.getMonth(date), this.getDate(date));
        };
        /**
         * @param {?} year
         * @param {?} month
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.createDate = function (year, month, date) {
            // Check for invalid month and date (except upper bound on date which we have to check after
            // creating the Date).
            if (month < 0 || month > 11) {
                throw Error("Invalid month index \"" + month + "\". Month index has to be between 0 and 11.");
            }
            if (date < 1) {
                throw Error("Invalid date \"" + date + "\". Date has to be greater than 0.");
            }
            var /** @type {?} */ result = this._createDateWithOverflow(year, month, date);
            // Check that the date wasn't above the upper bound for the month, causing the month to overflow
            if (result.getMonth() != month) {
                throw Error("Invalid date \"" + date + "\" for month with index \"" + month + "\".");
            }
            return result;
        };
        /**
         * @return {?}
         */
        NativeDateAdapter.prototype.today = function () {
            return new Date();
        };
        /**
         * @param {?} value
         * @return {?}
         */
        NativeDateAdapter.prototype.parse = function (value) {
            // We have no way using the native JS Date to set the parse format or locale, so we ignore these
            // parameters.
            if (typeof value == 'number') {
                return new Date(value);
            }
            return value ? new Date(Date.parse(value)) : null;
        };
        /**
         * @param {?} date
         * @param {?} displayFormat
         * @return {?}
         */
        NativeDateAdapter.prototype.format = function (date, displayFormat) {
            if (!this.isValid(date)) {
                throw Error('NativeDateAdapter: Cannot format invalid date.');
            }
            if (SUPPORTS_INTL_API) {
                // On IE and Edge the i18n API will throw a hard error that can crash the entire app
                // if we attempt to format a date whose year is less than 1 or greater than 9999.
                if (this._clampDate && (date.getFullYear() < 1 || date.getFullYear() > 9999)) {
                    date = this.clone(date);
                    date.setFullYear(Math.max(1, Math.min(9999, date.getFullYear())));
                }
                displayFormat = Object.assign({}, displayFormat, { timeZone: 'utc' });
                var /** @type {?} */ dtf = new Intl.DateTimeFormat(this.locale, displayFormat);
                return this._stripDirectionalityCharacters(this._format(dtf, date));
            }
            return this._stripDirectionalityCharacters(date.toDateString());
        };
        /**
         * @param {?} date
         * @param {?} years
         * @return {?}
         */
        NativeDateAdapter.prototype.addCalendarYears = function (date, years) {
            return this.addCalendarMonths(date, years * 12);
        };
        /**
         * @param {?} date
         * @param {?} months
         * @return {?}
         */
        NativeDateAdapter.prototype.addCalendarMonths = function (date, months) {
            var /** @type {?} */ newDate = this._createDateWithOverflow(this.getYear(date), this.getMonth(date) + months, this.getDate(date));
            // It's possible to wind up in the wrong month if the original month has more days than the new
            // month. In this case we want to go to the last day of the desired month.
            // Note: the additional + 12 % 12 ensures we end up with a positive number, since JS % doesn't
            // guarantee this.
            if (this.getMonth(newDate) != ((this.getMonth(date) + months) % 12 + 12) % 12) {
                newDate = this._createDateWithOverflow(this.getYear(newDate), this.getMonth(newDate), 0);
            }
            return newDate;
        };
        /**
         * @param {?} date
         * @param {?} days
         * @return {?}
         */
        NativeDateAdapter.prototype.addCalendarDays = function (date, days) {
            return this._createDateWithOverflow(this.getYear(date), this.getMonth(date), this.getDate(date) + days);
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.toIso8601 = function (date) {
            return [
                date.getUTCFullYear(),
                this._2digit(date.getUTCMonth() + 1),
                this._2digit(date.getUTCDate())
            ].join('-');
        };
        /**
         * Returns the given value if given a valid Date or null. Deserializes valid ISO 8601 strings
         * (https://www.ietf.org/rfc/rfc3339.txt) into valid Dates and empty string into null. Returns an
         * invalid date for all other values.
         * @param {?} value
         * @return {?}
         */
        NativeDateAdapter.prototype.deserialize = function (value) {
            if (typeof value === 'string') {
                if (!value) {
                    return null;
                }
                // The `Date` constructor accepts formats other than ISO 8601, so we need to make sure the
                // string is the right format first.
                if (ISO_8601_REGEX.test(value)) {
                    var /** @type {?} */ date = new Date(value);
                    if (this.isValid(date)) {
                        return date;
                    }
                }
            }
            return _super.prototype.deserialize.call(this, value);
        };
        /**
         * @param {?} obj
         * @return {?}
         */
        NativeDateAdapter.prototype.isDateInstance = function (obj) {
            return obj instanceof Date;
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype.isValid = function (date) {
            return !isNaN(date.getTime());
        };
        /**
         * @return {?}
         */
        NativeDateAdapter.prototype.invalid = function () {
            return new Date(NaN);
        };
        /**
         * Creates a date but allows the month and date to overflow.
         * @param {?} year
         * @param {?} month
         * @param {?} date
         * @return {?}
         */
        NativeDateAdapter.prototype._createDateWithOverflow = function (year, month, date) {
            var /** @type {?} */ result = new Date(year, month, date);
            // We need to correct for the fact that JS native Date treats years in range [0, 99] as
            // abbreviations for 19xx.
            if (year >= 0 && year < 100) {
                result.setFullYear(this.getYear(result) - 1900);
            }
            return result;
        };
        /**
         * Pads a number to make it two digits.
         * @param {?} n The number to pad.
         * @return {?} The padded number.
         */
        NativeDateAdapter.prototype._2digit = function (n) {
            return ('00' + n).slice(-2);
        };
        /**
         * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
         * other browsers do not. We remove them to make output consistent and because they interfere with
         * date parsing.
         * @param {?} str The string to strip direction characters from.
         * @return {?} The stripped string.
         */
        NativeDateAdapter.prototype._stripDirectionalityCharacters = function (str) {
            return str.replace(/[\u200e\u200f]/g, '');
        };
        /**
         * When converting Date object to string, javascript built-in functions may return wrong
         * results because it applies its internal DST rules. The DST rules around the world change
         * very frequently, and the current valid rule is not always valid in previous years though.
         * We work around this problem building a new Date object which has its internal UTC
         * representation with the local date and time.
         * @param {?} dtf Intl.DateTimeFormat object, containg the desired string format. It must have
         *    timeZone set to 'utc' to work fine.
         * @param {?} date Date from which we want to get the string representation according to dtf
         * @return {?} A Date object with its UTC representation based on the passed in date info
         */
        NativeDateAdapter.prototype._format = function (dtf, date) {
            var /** @type {?} */ d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds()));
            return dtf.format(d);
        };
        return NativeDateAdapter;
    }(DateAdapter));
    NativeDateAdapter.decorators = [
        { type: core.Injectable },
    ];
    /** @nocollapse */
    NativeDateAdapter.ctorParameters = function () {
        return [
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_LOCALE,] },] },
            { type: platform.Platform, },
        ];
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var /** @type {?} */ MAT_NATIVE_DATE_FORMATS = {
        parse: {
            dateInput: null,
        },
        display: {
            dateInput: { year: 'numeric', month: 'numeric', day: 'numeric' },
            monthYearLabel: { year: 'numeric', month: 'short' },
            dateA11yLabel: { year: 'numeric', month: 'long', day: 'numeric' },
            monthYearA11yLabel: { year: 'numeric', month: 'long' },
        }
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NativeDateModule = (function () {
        function NativeDateModule() {
        }
        return NativeDateModule;
    }());
    NativeDateModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [platform.PlatformModule],
                    providers: [
                        { provide: DateAdapter, useClass: NativeDateAdapter },
                    ],
                },] },
    ];
    var ɵ0$1 = MAT_NATIVE_DATE_FORMATS;
    var MatRangeNativeDateModule = (function () {
        function MatRangeNativeDateModule() {
        }
        return MatRangeNativeDateModule;
    }());
    MatRangeNativeDateModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [NativeDateModule],
                    providers: [{ provide: MAT_DATE_FORMATS, useValue: ɵ0$1 }],
                },] },
    ];
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /**
     * \@docs-private
     * @param {?} provider
     * @return {?}
     */
    function createMissingDateImplError(provider) {
        return Error("MatDatepicker: No provider found for " + provider + ". You must import one of the following " +
            "modules at your application root: MatNativeDateModule, MatMomentDateModule, or provide a " +
            "custom implementation.");
    }
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * Datepicker data that requires internationalization.
     */
    var matRangeDatepickerIntl = (function () {
        function matRangeDatepickerIntl() {
            /**
             * Stream that emits whenever the labels here are changed. Use this to notify
             * components if the labels have changed after initialization.
             */
            this.changes = new rxjs.Subject();
            /**
             * A label for the calendar popup (used by screen readers).
             */
            this.calendarLabel = 'Calendar';
            /**
             * A label for the button used to open the calendar popup (used by screen readers).
             */
            this.openCalendarLabel = 'Open calendar';
            /**
             * A label for the previous month button (used by screen readers).
             */
            this.prevMonthLabel = 'Previous month';
            /**
             * A label for the next month button (used by screen readers).
             */
            this.nextMonthLabel = 'Next month';
            /**
             * A label for the previous year button (used by screen readers).
             */
            this.prevYearLabel = 'Previous year';
            /**
             * A label for the next year button (used by screen readers).
             */
            this.nextYearLabel = 'Next year';
            /**
             * A label for the previous multi-year button (used by screen readers).
             */
            this.prevMultiYearLabel = 'Previous 20 years';
            /**
             * A label for the next multi-year button (used by screen readers).
             */
            this.nextMultiYearLabel = 'Next 20 years';
            /**
             * A label for the 'switch to month view' button (used by screen readers).
             */
            this.switchToMonthViewLabel = 'Choose date';
            /**
             * A label for the 'switch to year view' button (used by screen readers).
             */
            this.switchToMultiYearViewLabel = 'Choose month and year';
        }
        return matRangeDatepickerIntl;
    }());
    matRangeDatepickerIntl.decorators = [
        { type: core.Injectable, args: [{ providedIn: 'root' },] },
    ];
    /** @nocollapse */ matRangeDatepickerIntl.ngInjectableDef = core.defineInjectable({ factory: function matRangeDatepickerIntl_Factory() { return new matRangeDatepickerIntl(); }, token: matRangeDatepickerIntl, providedIn: "root" });
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * An internal class that represents the data corresponding to a single calendar cell.
     * \@docs-private
     */
    var SatCalendarCell = (function () {
        /**
         * @param {?} value
         * @param {?} displayValue
         * @param {?} ariaLabel
         * @param {?} enabled
         */
        function SatCalendarCell(value, displayValue, ariaLabel, enabled) {
            this.value = value;
            this.displayValue = displayValue;
            this.ariaLabel = ariaLabel;
            this.enabled = enabled;
        }
        return SatCalendarCell;
    }());
    /**
     * An internal component used to display calendar data in a table.
     * \@docs-private
     */
    var SatCalendarBody = (function () {
        /**
         * @param {?} _elementRef
         * @param {?} _ngZone
         */
        function SatCalendarBody(_elementRef, _ngZone) {
            this._elementRef = _elementRef;
            this._ngZone = _ngZone;
            /**
             * Whether to use date range selection behaviour.
             */
            this.rangeMode = false;
            /**
             * The number of columns in the table.
             */
            this.numCols = 7;
            /**
             * Whether to allow selection of disabled cells.
             */
            this.allowDisabledSelection = false;
            /**
             * The cell number of the active cell in the table.
             */
            this.activeCell = 0;
            /**
             * The aspect ratio (width / height) to use for the cells in the table. This aspect ratio will be
             * maintained even as the table resizes.
             */
            this.cellAspectRatio = 1;
            /**
             * Emits when a new value is selected.
             */
            this.selectedValueChange = new core.EventEmitter();
        }
        /**
         * @param {?} cell
         * @return {?}
         */
        SatCalendarBody.prototype._cellClicked = function (cell) {
            if (!this.allowDisabledSelection && !cell.enabled) {
                return;
            }
            this.selectedValueChange.emit(cell.value);
        };
        Object.defineProperty(SatCalendarBody.prototype, "_firstRowOffset", {
            /**
             * The number of blank cells to put at the beginning for the first row.
             * @return {?}
             */
            get: function () {
                return this.rows && this.rows.length && this.rows[0].length ?
                    this.numCols - this.rows[0].length : 0;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} rowIndex
         * @param {?} colIndex
         * @return {?}
         */
        SatCalendarBody.prototype._isActiveCell = function (rowIndex, colIndex) {
            var /** @type {?} */ cellNumber = rowIndex * this.numCols + colIndex;
            // Account for the fact that the first row may not have as many cells.
            if (rowIndex) {
                cellNumber -= this._firstRowOffset;
            }
            return cellNumber == this.activeCell;
        };
        /**
         * Whenever to mark cell as semi-selected (inside dates interval).
         * @param {?} date
         * @return {?}
         */
        SatCalendarBody.prototype._isSemiSelected = function (date) {
            if (!this.rangeMode) {
                return false;
            }
            if (this.rangeFull) {
                return true;
            }
            /** Do not mark start and end of interval. */
            if (date === this.begin || date === this.end) {
                return false;
            }
            if (this.begin && !this.end) {
                return date > this.begin;
            }
            if (this.end && !this.begin) {
                return date < this.end;
            }
            return date > /** @type {?} */ (this.begin) && date < /** @type {?} */ (this.end);
        };
        /**
         * Focuses the active cell after the microtask queue is empty.
         * @return {?}
         */
        SatCalendarBody.prototype._focusActiveCell = function () {
            var _this = this;
            this._ngZone.runOutsideAngular(function () {
                _this._ngZone.onStable.asObservable().pipe(operators.take(1)).subscribe(function () {
                    _this._elementRef.nativeElement.querySelector('.mat-calendar-body-active').focus();
                });
            });
        };
        return SatCalendarBody;
    }());
    SatCalendarBody.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: '[sat-calendar-body]',
                    template: "<!--\n  If there's not enough space in the first row, create a separate label row. We mark this row as\n  aria-hidden because we don't want it to be read out as one of the weeks in the month.\n-->\n<tr *ngIf=\"_firstRowOffset < labelMinRequiredCells\" aria-hidden=\"true\">\n  <td class=\"mat-calendar-body-label\"\n      [attr.colspan]=\"numCols\"\n      [style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"\n      [style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">\n    {{label}}\n  </td>\n</tr>\n\n<!-- Create the first row separately so we can include a special spacer cell. -->\n<tr *ngFor=\"let row of rows; let rowIndex = index\" role=\"row\">\n  <!--\n    We mark this cell as aria-hidden so it doesn't get read out as one of the days in the week.\n    The aspect ratio of the table cells is maintained by setting the top and bottom padding as a\n    percentage of the width (a variant of the trick described here:\n    https://www.w3schools.com/howto/howto_css_aspect_ratio.asp).\n  -->\n  <td *ngIf=\"rowIndex === 0 && _firstRowOffset\"\n      aria-hidden=\"true\"\n      class=\"mat-calendar-body-label\"\n      [attr.colspan]=\"_firstRowOffset\"\n      [style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"\n      [style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">\n    {{_firstRowOffset >= labelMinRequiredCells ? label : ''}}\n  </td>\n  <td *ngFor=\"let item of row; let colIndex = index\"\n      role=\"gridcell\"\n      class=\"mat-calendar-body-cell\"\n      [tabindex]=\"_isActiveCell(rowIndex, colIndex) ? 0 : -1\"\n      [class.mat-calendar-body-disabled]=\"!item.enabled\"\n      [class.mat-calendar-body-active]=\"_isActiveCell(rowIndex, colIndex)\"\n      [class.mat-calendar-body-begin-range]=\"begin === item.value\"\n      [class.mat-calendar-body-end-range]=\"end === item.value\"\n      [class.mat-calendar-cell-semi-selected]=\"_isSemiSelected(item.value)\"\n      [attr.aria-label]=\"item.ariaLabel\"\n      [attr.aria-disabled]=\"!item.enabled || null\"\n      (click)=\"_cellClicked(item)\"\n      [style.width.%]=\"100 / numCols\"\n      [style.paddingTop.%]=\"50 * cellAspectRatio / numCols\"\n      [style.paddingBottom.%]=\"50 * cellAspectRatio / numCols\">\n    <div class=\"mat-calendar-body-cell-content\"\n         [class.mat-calendar-body-selected]=\"begin === item.value || end === item.value || selectedValue === item.value\"\n         [class.mat-calendar-body-semi-selected]=\"_isSemiSelected(item.value)\"\n         [class.mat-calendar-body-today]=\"todayValue === item.value\">\n      {{item.displayValue}}\n    </div>\n  </td>\n</tr>\n",
                    styles: [".mat-calendar-body{min-width:224px}.mat-calendar-body-label{height:0;line-height:0;text-align:left;padding-left:4.71429%;padding-right:4.71429%}.mat-calendar-body-cell{position:relative;height:0;line-height:0;text-align:center;outline:0;cursor:pointer}.mat-calendar-body-disabled{cursor:default}.mat-calendar-body-cell-content{position:absolute;top:5%;left:5%;display:flex;align-items:center;justify-content:center;box-sizing:border-box;width:90%;height:90%;line-height:1;border-width:1px;border-style:solid;border-radius:999px}[dir=rtl] .mat-calendar-body-label{text-align:right}"],
                    host: {
                        'class': 'mat-calendar-body',
                        'role': 'grid',
                        'attr.aria-readonly': 'true'
                    },
                    exportAs: 'matCalendarBody',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    SatCalendarBody.ctorParameters = function () {
        return [
            { type: core.ElementRef, },
            { type: core.NgZone, },
        ];
    };
    SatCalendarBody.propDecorators = {
        "label": [{ type: core.Input },],
        "rows": [{ type: core.Input },],
        "todayValue": [{ type: core.Input },],
        "selectedValue": [{ type: core.Input },],
        "begin": [{ type: core.Input },],
        "end": [{ type: core.Input },],
        "rangeFull": [{ type: core.Input },],
        "rangeMode": [{ type: core.Input },],
        "labelMinRequiredCells": [{ type: core.Input },],
        "numCols": [{ type: core.Input },],
        "allowDisabledSelection": [{ type: core.Input },],
        "activeCell": [{ type: core.Input },],
        "cellAspectRatio": [{ type: core.Input },],
        "selectedValueChange": [{ type: core.Output },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ DAYS_PER_WEEK = 7;
    /**
     * An internal component used to display a single month in the datepicker.
     * \@docs-private
     * @template D
     */
    var SatMonthView = (function () {
        /**
         * @param {?} _changeDetectorRef
         * @param {?} _dateFormats
         * @param {?} _dateAdapter
         * @param {?=} _dir
         */
        function SatMonthView(_changeDetectorRef, _dateFormats, _dateAdapter, _dir) {
            this._changeDetectorRef = _changeDetectorRef;
            this._dateFormats = _dateFormats;
            this._dateAdapter = _dateAdapter;
            this._dir = _dir;
            /**
             * Allow selecting range of dates.
             */
            this.rangeMode = false;
            /**
             * Whenever full month is inside dates interval.
             */
            this._rangeFull = false;
            /**
             * Whenever user already selected start of dates interval.
             */
            this._beginDateSelected = false;
            /**
             * Emits when a new date is selected.
             */
            this.selectedChange = new core.EventEmitter();
            /**
             * Emits when any date is selected.
             */
            this._userSelection = new core.EventEmitter();
            /**
             * Emits when any date is activated.
             */
            this.activeDateChange = new core.EventEmitter();
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            if (!this._dateFormats) {
                throw createMissingDateImplError('MAT_DATE_FORMATS');
            }
            var /** @type {?} */ firstDayOfWeek = this._dateAdapter.getFirstDayOfWeek();
            var /** @type {?} */ narrowWeekdays = this._dateAdapter.getDayOfWeekNames('narrow');
            var /** @type {?} */ longWeekdays = this._dateAdapter.getDayOfWeekNames('long');
            // Rotate the labels for days of the week based on the configured first day of the week.
            var /** @type {?} */ weekdays = longWeekdays.map(function (long, i) {
                return { long: long, narrow: narrowWeekdays[i] };
            });
            this._weekdays = weekdays.slice(firstDayOfWeek).concat(weekdays.slice(0, firstDayOfWeek));
            this._activeDate = this._dateAdapter.today();
        }
        Object.defineProperty(SatMonthView.prototype, "beginDate", {
            /**
             * Current start of interval.
             * @return {?}
             */
            get: function () { return this._beginDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._beginDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this.updateRangeSpecificValues();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMonthView.prototype, "endDate", {
            /**
             * Current end of interval.
             * @return {?}
             */
            get: function () { return this._endDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._endDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this.updateRangeSpecificValues();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMonthView.prototype, "activeDate", {
            /**
             * The date to display in this month view (everything other than the month and year is ignored).
             * @return {?}
             */
            get: function () { return this._activeDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                var /** @type {?} */ oldActiveDate = this._activeDate;
                var /** @type {?} */ validDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
                this._activeDate = this._dateAdapter.clampDate(validDate, this.minDate, this.maxDate);
                if (!this._hasSameMonthAndYear(oldActiveDate, this._activeDate)) {
                    this._init();
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMonthView.prototype, "selected", {
            /**
             * The currently selected date.
             * @return {?}
             */
            get: function () { return this._selected; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._selected = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this._selectedDate = this._getDateInCurrentMonth(this._selected);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMonthView.prototype, "minDate", {
            /**
             * The minimum selectable date.
             * @return {?}
             */
            get: function () { return this._minDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._minDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMonthView.prototype, "maxDate", {
            /**
             * The maximum selectable date.
             * @return {?}
             */
            get: function () { return this._maxDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._maxDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SatMonthView.prototype.ngAfterContentInit = function () {
            this._init();
        };
        /**
         * Handles when a new date is selected.
         * @param {?} date
         * @return {?}
         */
        SatMonthView.prototype._dateSelected = function (date) {
            if (this.rangeMode) {
                var /** @type {?} */ selectedYear = this._dateAdapter.getYear(this.activeDate);
                var /** @type {?} */ selectedMonth = this._dateAdapter.getMonth(this.activeDate);
                var /** @type {?} */ selectedDate = this._dateAdapter.createDate(selectedYear, selectedMonth, date);
                if (!this._beginDateSelected) {
                    // At first click emit the same start and end of interval
                    this._beginDateSelected = true;
                    this.selectedChange.emit(selectedDate);
                }
                else {
                    this._beginDateSelected = false;
                    this.selectedChange.emit(selectedDate);
                    this._userSelection.emit();
                }
            }
            else if (this._selectedDate != date) {
                var /** @type {?} */ selectedYear = this._dateAdapter.getYear(this.activeDate);
                var /** @type {?} */ selectedMonth = this._dateAdapter.getMonth(this.activeDate);
                var /** @type {?} */ selectedDate = this._dateAdapter.createDate(selectedYear, selectedMonth, date);
                this.selectedChange.emit(selectedDate);
                this._userSelection.emit();
            }
        };
        /**
         * Handles keydown events on the calendar body when calendar is in month view.
         * @param {?} event
         * @return {?}
         */
        SatMonthView.prototype._handleCalendarBodyKeydown = function (event) {
            // TODO(mmalerba): We currently allow keyboard navigation to disabled dates, but just prevent
            // disabled ones from being selected. This may not be ideal, we should look into whether
            // navigation should skip over disabled dates, and if so, how to implement that efficiently.
            var /** @type {?} */ oldActiveDate = this._activeDate;
            var /** @type {?} */ isRtl = this._isRtl();
            switch (event.keyCode) {
                case keycodes.LEFT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, isRtl ? 1 : -1);
                    break;
                case keycodes.RIGHT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, isRtl ? -1 : 1);
                    break;
                case keycodes.UP_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, -7);
                    break;
                case keycodes.DOWN_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, 7);
                    break;
                case keycodes.HOME:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, 1 - this._dateAdapter.getDate(this._activeDate));
                    break;
                case keycodes.END:
                    this.activeDate = this._dateAdapter.addCalendarDays(this._activeDate, (this._dateAdapter.getNumDaysInMonth(this._activeDate) -
                        this._dateAdapter.getDate(this._activeDate)));
                    break;
                case keycodes.PAGE_UP:
                    this.activeDate = event.altKey ?
                        this._dateAdapter.addCalendarYears(this._activeDate, -1) :
                        this._dateAdapter.addCalendarMonths(this._activeDate, -1);
                    break;
                case keycodes.PAGE_DOWN:
                    this.activeDate = event.altKey ?
                        this._dateAdapter.addCalendarYears(this._activeDate, 1) :
                        this._dateAdapter.addCalendarMonths(this._activeDate, 1);
                    break;
                case keycodes.ENTER:
                    if (!this.dateFilter || this.dateFilter(this._activeDate)) {
                        this._dateSelected(this._dateAdapter.getDate(this._activeDate));
                        this._userSelection.emit();
                        // Prevent unexpected default actions such as form submission.
                        event.preventDefault();
                    }
                    return;
                default:
                    // Don't prevent default or focus active cell on keys that we don't explicitly handle.
                    return;
            }
            if (this._dateAdapter.compareDate(oldActiveDate, this.activeDate)) {
                this.activeDateChange.emit(this.activeDate);
            }
            this._focusActiveCell();
            // Prevent unexpected default actions such as form submission.
            event.preventDefault();
        };
        /**
         * Initializes this month view.
         * @return {?}
         */
        SatMonthView.prototype._init = function () {
            this.updateRangeSpecificValues();
            this._selectedDate = this._getDateInCurrentMonth(this.selected);
            this._todayDate = this._getDateInCurrentMonth(this._dateAdapter.today());
            this._monthLabel =
                this._dateAdapter.getMonthNames('short')[this._dateAdapter.getMonth(this.activeDate)]
                    .toLocaleUpperCase();
            var /** @type {?} */ firstOfMonth = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), this._dateAdapter.getMonth(this.activeDate), 1);
            this._firstWeekOffset =
                (DAYS_PER_WEEK + this._dateAdapter.getDayOfWeek(firstOfMonth) -
                    this._dateAdapter.getFirstDayOfWeek()) % DAYS_PER_WEEK;
            this._createWeekCells();
            this._changeDetectorRef.markForCheck();
        };
        /**
         * Focuses the active cell after the microtask queue is empty.
         * @return {?}
         */
        SatMonthView.prototype._focusActiveCell = function () {
            this._matCalendarBody._focusActiveCell();
        };
        /**
         * Creates SatCalendarCells for the dates in this month.
         * @return {?}
         */
        SatMonthView.prototype._createWeekCells = function () {
            var /** @type {?} */ daysInMonth = this._dateAdapter.getNumDaysInMonth(this.activeDate);
            var /** @type {?} */ dateNames = this._dateAdapter.getDateNames();
            this._weeks = [[]];
            for (var /** @type {?} */ i = 0, /** @type {?} */ cell = this._firstWeekOffset; i < daysInMonth; i++, cell++) {
                if (cell == DAYS_PER_WEEK) {
                    this._weeks.push([]);
                    cell = 0;
                }
                var /** @type {?} */ date = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), this._dateAdapter.getMonth(this.activeDate), i + 1);
                var /** @type {?} */ enabled = this._shouldEnableDate(date);
                var /** @type {?} */ ariaLabel = this._dateAdapter.format(date, this._dateFormats.display.dateA11yLabel);
                this._weeks[this._weeks.length - 1]
                    .push(new SatCalendarCell(i + 1, dateNames[i], ariaLabel, enabled));
            }
        };
        /**
         * Date filter for the month
         * @param {?} date
         * @return {?}
         */
        SatMonthView.prototype._shouldEnableDate = function (date) {
            return !!date &&
                (!this.dateFilter || this.dateFilter(date)) &&
                (!this.minDate || this._dateAdapter.compareDate(date, this.minDate) >= 0) &&
                (!this.maxDate || this._dateAdapter.compareDate(date, this.maxDate) <= 0);
        };
        /**
         * Gets the date in this month that the given Date falls on.
         * Returns null if the given Date is in another month.
         * @param {?} date
         * @return {?}
         */
        SatMonthView.prototype._getDateInCurrentMonth = function (date) {
            return date && this._hasSameMonthAndYear(date, this.activeDate) ?
                this._dateAdapter.getDate(date) : null;
        };
        /**
         * Checks whether the 2 dates are non-null and fall within the same month of the same year.
         * @param {?} d1
         * @param {?} d2
         * @return {?}
         */
        SatMonthView.prototype._hasSameMonthAndYear = function (d1, d2) {
            return !!(d1 && d2 && this._dateAdapter.getMonth(d1) == this._dateAdapter.getMonth(d2) &&
                this._dateAdapter.getYear(d1) == this._dateAdapter.getYear(d2));
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        SatMonthView.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        /**
         * Determines whether the user has the RTL layout direction.
         * @return {?}
         */
        SatMonthView.prototype._isRtl = function () {
            return this._dir && this._dir.value === 'rtl';
        };
        /**
         * Updates range full parameter on each begin or end of interval update.
         * Necessary to display calendar-body correctly
         * @return {?}
         */
        SatMonthView.prototype.updateRangeSpecificValues = function () {
            if (this.rangeMode) {
                this._beginDateNumber = this._getDateInCurrentMonth(this._beginDate);
                this._endDateNumber = this._getDateInCurrentMonth(this._endDate);
                this._rangeFull = this.beginDate && this.endDate && !this._beginDateNumber &&
                    !this._endDateNumber &&
                    this._dateAdapter.compareDate(this.beginDate, this.activeDate) <= 0 &&
                    this._dateAdapter.compareDate(this.activeDate, this.endDate) <= 0;
            }
            else {
                this._beginDateNumber = this._endDateNumber = null;
                this._rangeFull = false;
            }
        };
        return SatMonthView;
    }());
    SatMonthView.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'sat-month-view',
                    template: "<table class=\"mat-calendar-table\">\n  <thead class=\"mat-calendar-table-header\">\n    <tr><th *ngFor=\"let day of _weekdays\" [attr.aria-label]=\"day.long\">{{day.narrow}}</th></tr>\n    <tr><th class=\"mat-calendar-table-header-divider\" colspan=\"7\" aria-hidden=\"true\"></th></tr>\n  </thead>\n  <tbody sat-calendar-body\n         [label]=\"_monthLabel\"\n         [rows]=\"_weeks\"\n         [todayValue]=\"_todayDate\"\n         [selectedValue]=\"_selectedDate\"\n         [begin]=\"_beginDateNumber\"\n         [end]=\"_endDateNumber\"\n         [rangeFull]=\"_rangeFull\"\n         [rangeMode]=\"rangeMode\"\n         [labelMinRequiredCells]=\"3\"\n         [activeCell]=\"_dateAdapter.getDate(activeDate) - 1\"\n         (selectedValueChange)=\"_dateSelected($event)\"\n         (keydown)=\"_handleCalendarBodyKeydown($event)\">\n  </tbody>\n</table>\n",
                    exportAs: 'matMonthView',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    SatMonthView.ctorParameters = function () {
        return [
            { type: core.ChangeDetectorRef, },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: bidi.Directionality, decorators: [{ type: core.Optional },] },
        ];
    };
    SatMonthView.propDecorators = {
        "beginDate": [{ type: core.Input },],
        "endDate": [{ type: core.Input },],
        "rangeMode": [{ type: core.Input },],
        "activeDate": [{ type: core.Input },],
        "selected": [{ type: core.Input },],
        "minDate": [{ type: core.Input },],
        "maxDate": [{ type: core.Input },],
        "dateFilter": [{ type: core.Input },],
        "selectedChange": [{ type: core.Output },],
        "_userSelection": [{ type: core.Output },],
        "activeDateChange": [{ type: core.Output },],
        "_matCalendarBody": [{ type: core.ViewChild, args: [SatCalendarBody,] },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ yearsPerPage = 24;
    var /** @type {?} */ yearsPerRow = 4;
    /**
     * An internal component used to display a year selector in the datepicker.
     * \@docs-private
     * @template D
     */
    var SatMultiYearView = (function () {
        /**
         * @param {?} _changeDetectorRef
         * @param {?} _dateAdapter
         * @param {?=} _dir
         */
        function SatMultiYearView(_changeDetectorRef, _dateAdapter, _dir) {
            this._changeDetectorRef = _changeDetectorRef;
            this._dateAdapter = _dateAdapter;
            this._dir = _dir;
            /**
             * Emits when a new year is selected.
             */
            this.selectedChange = new core.EventEmitter();
            /**
             * Emits the selected year. This doesn't imply a change on the selected date
             */
            this.yearSelected = new core.EventEmitter();
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            this._activeDate = this._dateAdapter.today();
        }
        Object.defineProperty(SatMultiYearView.prototype, "activeDate", {
            /**
             * The date to display in this multi-year view (everything other than the year is ignored).
             * @return {?}
             */
            get: function () { return this._activeDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                var /** @type {?} */ oldActiveDate = this._activeDate;
                var /** @type {?} */ validDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
                this._activeDate = this._dateAdapter.clampDate(validDate, this.minDate, this.maxDate);
                if (Math.floor(this._dateAdapter.getYear(oldActiveDate) / yearsPerPage) !=
                    Math.floor(this._dateAdapter.getYear(this._activeDate) / yearsPerPage)) {
                    this._init();
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMultiYearView.prototype, "selected", {
            /**
             * The currently selected date.
             * @return {?}
             */
            get: function () { return this._selected; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._selected = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this._selectedYear = this._selected && this._dateAdapter.getYear(this._selected);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMultiYearView.prototype, "minDate", {
            /**
             * The minimum selectable date.
             * @return {?}
             */
            get: function () { return this._minDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._minDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatMultiYearView.prototype, "maxDate", {
            /**
             * The maximum selectable date.
             * @return {?}
             */
            get: function () { return this._maxDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._maxDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SatMultiYearView.prototype.ngAfterContentInit = function () {
            this._init();
        };
        /**
         * Initializes this multi-year view.
         * @return {?}
         */
        SatMultiYearView.prototype._init = function () {
            var _this = this;
            this._todayYear = this._dateAdapter.getYear(this._dateAdapter.today());
            var /** @type {?} */ activeYear = this._dateAdapter.getYear(this._activeDate);
            var /** @type {?} */ activeOffset = activeYear % yearsPerPage;
            this._years = [];
            for (var /** @type {?} */ i = 0, /** @type {?} */ row = []; i < yearsPerPage; i++) {
                row.push(activeYear - activeOffset + i);
                if (row.length == yearsPerRow) {
                    this._years.push(row.map(function (year) { return _this._createCellForYear(year); }));
                    row = [];
                }
            }
            this._changeDetectorRef.markForCheck();
        };
        /**
         * Handles when a new year is selected.
         * @param {?} year
         * @return {?}
         */
        SatMultiYearView.prototype._yearSelected = function (year) {
            this.yearSelected.emit(this._dateAdapter.createDate(year, 0, 1));
            var /** @type {?} */ month = this._dateAdapter.getMonth(this.activeDate);
            var /** @type {?} */ daysInMonth = this._dateAdapter.getNumDaysInMonth(this._dateAdapter.createDate(year, month, 1));
            this.selectedChange.emit(this._dateAdapter.createDate(year, month, Math.min(this._dateAdapter.getDate(this.activeDate), daysInMonth)));
        };
        /**
         * Handles keydown events on the calendar body when calendar is in multi-year view.
         * @param {?} event
         * @return {?}
         */
        SatMultiYearView.prototype._handleCalendarBodyKeydown = function (event) {
            // TODO(mmalerba): We currently allow keyboard navigation to disabled dates, but just prevent
            // disabled ones from being selected. This may not be ideal, we should look into whether
            // navigation should skip over disabled dates, and if so, how to implement that efficiently.
            var /** @type {?} */ isRtl = this._isRtl();
            switch (event.keyCode) {
                case keycodes.LEFT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, isRtl ? 1 : -1);
                    break;
                case keycodes.RIGHT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, isRtl ? -1 : 1);
                    break;
                case keycodes.UP_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, -yearsPerRow);
                    break;
                case keycodes.DOWN_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, yearsPerRow);
                    break;
                case keycodes.HOME:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, -this._dateAdapter.getYear(this._activeDate) % yearsPerPage);
                    break;
                case keycodes.END:
                    this.activeDate = this._dateAdapter.addCalendarYears(this._activeDate, yearsPerPage - this._dateAdapter.getYear(this._activeDate) % yearsPerPage - 1);
                    break;
                case keycodes.PAGE_UP:
                    this.activeDate =
                        this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? -yearsPerPage * 10 : -yearsPerPage);
                    break;
                case keycodes.PAGE_DOWN:
                    this.activeDate =
                        this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? yearsPerPage * 10 : yearsPerPage);
                    break;
                case keycodes.ENTER:
                    this._yearSelected(this._dateAdapter.getYear(this._activeDate));
                    break;
                default:
                    // Don't prevent default or focus active cell on keys that we don't explicitly handle.
                    return;
            }
            this._focusActiveCell();
            // Prevent unexpected default actions such as form submission.
            event.preventDefault();
        };
        /**
         * @return {?}
         */
        SatMultiYearView.prototype._getActiveCell = function () {
            return this._dateAdapter.getYear(this.activeDate) % yearsPerPage;
        };
        /**
         * Focuses the active cell after the microtask queue is empty.
         * @return {?}
         */
        SatMultiYearView.prototype._focusActiveCell = function () {
            this._matCalendarBody._focusActiveCell();
        };
        /**
         * Creates an SatCalendarCell for the given year.
         * @param {?} year
         * @return {?}
         */
        SatMultiYearView.prototype._createCellForYear = function (year) {
            var /** @type {?} */ yearName = this._dateAdapter.getYearName(this._dateAdapter.createDate(year, 0, 1));
            return new SatCalendarCell(year, yearName, yearName, this._shouldEnableYear(year));
        };
        /**
         * Whether the given year is enabled.
         * @param {?} year
         * @return {?}
         */
        SatMultiYearView.prototype._shouldEnableYear = function (year) {
            // disable if the year is greater than maxDate lower than minDate
            if (year === undefined || year === null ||
                (this.maxDate && year > this._dateAdapter.getYear(this.maxDate)) ||
                (this.minDate && year < this._dateAdapter.getYear(this.minDate))) {
                return false;
            }
            // enable if it reaches here and there's no filter defined
            if (!this.dateFilter) {
                return true;
            }
            var /** @type {?} */ firstOfYear = this._dateAdapter.createDate(year, 0, 1);
            // If any date in the year is enabled count the year as enabled.
            for (var /** @type {?} */ date = firstOfYear; this._dateAdapter.getYear(date) == year; date = this._dateAdapter.addCalendarDays(date, 1)) {
                if (this.dateFilter(date)) {
                    return true;
                }
            }
            return false;
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        SatMultiYearView.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        /**
         * Determines whether the user has the RTL layout direction.
         * @return {?}
         */
        SatMultiYearView.prototype._isRtl = function () {
            return this._dir && this._dir.value === 'rtl';
        };
        return SatMultiYearView;
    }());
    SatMultiYearView.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'sat-multi-year-view',
                    template: "<table class=\"mat-calendar-table\">\n    <thead class=\"mat-calendar-table-header\">\n    <tr>\n        <th class=\"mat-calendar-table-header-divider\" colspan=\"4\"></th>\n    </tr>\n    </thead>\n    <tbody sat-calendar-body\n           allowDisabledSelection=\"true\"\n           [rows]=\"_years\"\n           [todayValue]=\"_todayYear\"\n           [selectedValue]=\"_selectedYear\"\n           [numCols]=\"4\"\n           [cellAspectRatio]=\"4 / 7\"\n           [activeCell]=\"_getActiveCell()\"\n           (selectedValueChange)=\"_yearSelected($event)\">\n    </tbody>\n</table>\n",
                    exportAs: 'matMultiYearView',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    SatMultiYearView.ctorParameters = function () {
        return [
            { type: core.ChangeDetectorRef, },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: bidi.Directionality, decorators: [{ type: core.Optional },] },
        ];
    };
    SatMultiYearView.propDecorators = {
        "activeDate": [{ type: core.Input },],
        "selected": [{ type: core.Input },],
        "minDate": [{ type: core.Input },],
        "maxDate": [{ type: core.Input },],
        "dateFilter": [{ type: core.Input },],
        "selectedChange": [{ type: core.Output },],
        "yearSelected": [{ type: core.Output },],
        "_matCalendarBody": [{ type: core.ViewChild, args: [SatCalendarBody,] },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * An internal component used to display a single year in the datepicker.
     * \@docs-private
     * @template D
     */
    var SatYearView = (function () {
        /**
         * @param {?} _changeDetectorRef
         * @param {?} _dateFormats
         * @param {?} _dateAdapter
         * @param {?=} _dir
         */
        function SatYearView(_changeDetectorRef, _dateFormats, _dateAdapter, _dir) {
            this._changeDetectorRef = _changeDetectorRef;
            this._dateFormats = _dateFormats;
            this._dateAdapter = _dateAdapter;
            this._dir = _dir;
            /**
             * Emits when a new month is selected.
             */
            this.selectedChange = new core.EventEmitter();
            /**
             * Emits the selected month. This doesn't imply a change on the selected date
             */
            this.monthSelected = new core.EventEmitter();
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            if (!this._dateFormats) {
                throw createMissingDateImplError('MAT_DATE_FORMATS');
            }
            this._activeDate = this._dateAdapter.today();
        }
        Object.defineProperty(SatYearView.prototype, "activeDate", {
            /**
             * The date to display in this year view (everything other than the year is ignored).
             * @return {?}
             */
            get: function () { return this._activeDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                var /** @type {?} */ oldActiveDate = this._activeDate;
                var /** @type {?} */ validDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value)) || this._dateAdapter.today();
                this._activeDate = this._dateAdapter.clampDate(validDate, this.minDate, this.maxDate);
                if (this._dateAdapter.getYear(oldActiveDate) !== this._dateAdapter.getYear(this._activeDate)) {
                    this._init();
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatYearView.prototype, "selected", {
            /**
             * The currently selected date.
             * @return {?}
             */
            get: function () { return this._selected; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._selected = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this._selectedMonth = this._getMonthInCurrentYear(this._selected);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatYearView.prototype, "minDate", {
            /**
             * The minimum selectable date.
             * @return {?}
             */
            get: function () { return this._minDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._minDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatYearView.prototype, "maxDate", {
            /**
             * The maximum selectable date.
             * @return {?}
             */
            get: function () { return this._maxDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._maxDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SatYearView.prototype.ngAfterContentInit = function () {
            this._init();
        };
        /**
         * Handles when a new month is selected.
         * @param {?} month
         * @return {?}
         */
        SatYearView.prototype._monthSelected = function (month) {
            var /** @type {?} */ normalizedDate = this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, 1);
            this.monthSelected.emit(normalizedDate);
            var /** @type {?} */ daysInMonth = this._dateAdapter.getNumDaysInMonth(normalizedDate);
            this.selectedChange.emit(this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, Math.min(this._dateAdapter.getDate(this.activeDate), daysInMonth)));
        };
        /**
         * Handles keydown events on the calendar body when calendar is in year view.
         * @param {?} event
         * @return {?}
         */
        SatYearView.prototype._handleCalendarBodyKeydown = function (event) {
            // TODO(mmalerba): We currently allow keyboard navigation to disabled dates, but just prevent
            // disabled ones from being selected. This may not be ideal, we should look into whether
            // navigation should skip over disabled dates, and if so, how to implement that efficiently.
            var /** @type {?} */ isRtl = this._isRtl();
            switch (event.keyCode) {
                case keycodes.LEFT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, isRtl ? 1 : -1);
                    break;
                case keycodes.RIGHT_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, isRtl ? -1 : 1);
                    break;
                case keycodes.UP_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, -4);
                    break;
                case keycodes.DOWN_ARROW:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, 4);
                    break;
                case keycodes.HOME:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, -this._dateAdapter.getMonth(this._activeDate));
                    break;
                case keycodes.END:
                    this.activeDate = this._dateAdapter.addCalendarMonths(this._activeDate, 11 - this._dateAdapter.getMonth(this._activeDate));
                    break;
                case keycodes.PAGE_UP:
                    this.activeDate =
                        this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? -10 : -1);
                    break;
                case keycodes.PAGE_DOWN:
                    this.activeDate =
                        this._dateAdapter.addCalendarYears(this._activeDate, event.altKey ? 10 : 1);
                    break;
                case keycodes.ENTER:
                    this._monthSelected(this._dateAdapter.getMonth(this._activeDate));
                    break;
                default:
                    // Don't prevent default or focus active cell on keys that we don't explicitly handle.
                    return;
            }
            this._focusActiveCell();
            // Prevent unexpected default actions such as form submission.
            event.preventDefault();
        };
        /**
         * Initializes this year view.
         * @return {?}
         */
        SatYearView.prototype._init = function () {
            var _this = this;
            this._selectedMonth = this._getMonthInCurrentYear(this.selected);
            this._todayMonth = this._getMonthInCurrentYear(this._dateAdapter.today());
            this._yearLabel = this._dateAdapter.getYearName(this.activeDate);
            var /** @type {?} */ monthNames = this._dateAdapter.getMonthNames('short');
            // First row of months only contains 5 elements so we can fit the year label on the same row.
            this._months = [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]].map(function (row) { return row.map(function (month) { return _this._createCellForMonth(month, monthNames[month]); }); });
            this._changeDetectorRef.markForCheck();
        };
        /**
         * Focuses the active cell after the microtask queue is empty.
         * @return {?}
         */
        SatYearView.prototype._focusActiveCell = function () {
            this._matCalendarBody._focusActiveCell();
        };
        /**
         * Gets the month in this year that the given Date falls on.
         * Returns null if the given Date is in another year.
         * @param {?} date
         * @return {?}
         */
        SatYearView.prototype._getMonthInCurrentYear = function (date) {
            return date && this._dateAdapter.getYear(date) == this._dateAdapter.getYear(this.activeDate) ?
                this._dateAdapter.getMonth(date) : null;
        };
        /**
         * Creates an SatCalendarCell for the given month.
         * @param {?} month
         * @param {?} monthName
         * @return {?}
         */
        SatYearView.prototype._createCellForMonth = function (month, monthName) {
            var /** @type {?} */ ariaLabel = this._dateAdapter.format(this._dateAdapter.createDate(this._dateAdapter.getYear(this.activeDate), month, 1), this._dateFormats.display.monthYearA11yLabel);
            return new SatCalendarCell(month, monthName.toLocaleUpperCase(), ariaLabel, this._shouldEnableMonth(month));
        };
        /**
         * Whether the given month is enabled.
         * @param {?} month
         * @return {?}
         */
        SatYearView.prototype._shouldEnableMonth = function (month) {
            var /** @type {?} */ activeYear = this._dateAdapter.getYear(this.activeDate);
            if (month === undefined || month === null ||
                this._isYearAndMonthAfterMaxDate(activeYear, month) ||
                this._isYearAndMonthBeforeMinDate(activeYear, month)) {
                return false;
            }
            if (!this.dateFilter) {
                return true;
            }
            var /** @type {?} */ firstOfMonth = this._dateAdapter.createDate(activeYear, month, 1);
            // If any date in the month is enabled count the month as enabled.
            for (var /** @type {?} */ date = firstOfMonth; this._dateAdapter.getMonth(date) == month; date = this._dateAdapter.addCalendarDays(date, 1)) {
                if (this.dateFilter(date)) {
                    return true;
                }
            }
            return false;
        };
        /**
         * Tests whether the combination month/year is after this.maxDate, considering
         * just the month and year of this.maxDate
         * @param {?} year
         * @param {?} month
         * @return {?}
         */
        SatYearView.prototype._isYearAndMonthAfterMaxDate = function (year, month) {
            if (this.maxDate) {
                var /** @type {?} */ maxYear = this._dateAdapter.getYear(this.maxDate);
                var /** @type {?} */ maxMonth = this._dateAdapter.getMonth(this.maxDate);
                return year > maxYear || (year === maxYear && month > maxMonth);
            }
            return false;
        };
        /**
         * Tests whether the combination month/year is before this.minDate, considering
         * just the month and year of this.minDate
         * @param {?} year
         * @param {?} month
         * @return {?}
         */
        SatYearView.prototype._isYearAndMonthBeforeMinDate = function (year, month) {
            if (this.minDate) {
                var /** @type {?} */ minYear = this._dateAdapter.getYear(this.minDate);
                var /** @type {?} */ minMonth = this._dateAdapter.getMonth(this.minDate);
                return year < minYear || (year === minYear && month < minMonth);
            }
            return false;
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        SatYearView.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        /**
         * Determines whether the user has the RTL layout direction.
         * @return {?}
         */
        SatYearView.prototype._isRtl = function () {
            return this._dir && this._dir.value === 'rtl';
        };
        return SatYearView;
    }());
    SatYearView.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'sat-year-view',
                    template: "<table class=\"mat-calendar-table\">\n  <thead class=\"mat-calendar-table-header\">\n  <tr><th class=\"mat-calendar-table-header-divider\" colspan=\"4\"></th></tr>\n  </thead>\n  <tbody sat-calendar-body\n         allowDisabledSelection=\"true\"\n         [label]=\"_yearLabel\"\n         [rows]=\"_months\"\n         [todayValue]=\"_todayMonth\"\n         [selectedValue]=\"_selectedMonth\"\n         [labelMinRequiredCells]=\"2\"\n         [numCols]=\"4\"\n         [cellAspectRatio]=\"4 / 7\"\n         [activeCell]=\"_dateAdapter.getMonth(activeDate)\"\n         (selectedValueChange)=\"_monthSelected($event)\"\n         (keydown)=\"_handleCalendarBodyKeydown($event)\">\n  </tbody>\n</table>\n",
                    exportAs: 'matYearView',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    SatYearView.ctorParameters = function () {
        return [
            { type: core.ChangeDetectorRef, },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: bidi.Directionality, decorators: [{ type: core.Optional },] },
        ];
    };
    SatYearView.propDecorators = {
        "activeDate": [{ type: core.Input },],
        "selected": [{ type: core.Input },],
        "minDate": [{ type: core.Input },],
        "maxDate": [{ type: core.Input },],
        "dateFilter": [{ type: core.Input },],
        "selectedChange": [{ type: core.Output },],
        "monthSelected": [{ type: core.Output },],
        "_matCalendarBody": [{ type: core.ViewChild, args: [SatCalendarBody,] },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * A calendar that is used as part of the datepicker.
     * \@docs-private
     * @template D
     */
    var SatCalendar = (function () {
        /**
         * @param {?} _intl
         * @param {?} _dateAdapter
         * @param {?} _dateFormats
         * @param {?} changeDetectorRef
         */
        function SatCalendar(_intl, _dateAdapter, _dateFormats, changeDetectorRef) {
            var _this = this;
            this._dateAdapter = _dateAdapter;
            this._dateFormats = _dateFormats;
            /**
             * Whenever datepicker is for selecting range of dates.
             */
            this.rangeMode = false;
            /**
             * Emits when new pair of dates selected.
             */
            this.dateSelected = new core.EventEmitter();
            /**
             * Whenever user already selected start of dates interval.
             */
            this._beginDateSelected = false;
            /**
             * Used for scheduling that focus should be moved to the active cell on the next tick.
             * We need to schedule it, rather than do it immediately, because we have to wait
             * for Angular to re-evaluate the view children.
             */
            this._moveFocusOnNextTick = false;
            /**
             * Whether the calendar should be started in month or year view.
             */
            this.startView = 'month';
            /**
             * Emits when the currently selected date changes.
             */
            this.selectedChange = new core.EventEmitter();
            /**
             * Emits the year chosen in multiyear view.
             * This doesn't imply a change on the selected date.
             */
            this.yearSelected = new core.EventEmitter();
            /**
             * Emits the month chosen in year view.
             * This doesn't imply a change on the selected date.
             */
            this.monthSelected = new core.EventEmitter();
            /**
             * Emits when any date is selected.
             */
            this._userSelection = new core.EventEmitter();
            /**
             * Emits whenever there is a state change that the header may need to respond to.
             */
            this.stateChanges = new rxjs.Subject();
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            if (!this._dateFormats) {
                throw createMissingDateImplError('MAT_DATE_FORMATS');
            }
            this._intlChanges = _intl.changes.subscribe(function () {
                changeDetectorRef.markForCheck();
                _this.stateChanges.next();
            });
        }
        Object.defineProperty(SatCalendar.prototype, "beginDate", {
            /**
             * Beginning of date range.
             * @return {?}
             */
            get: function () { return this._beginDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._beginDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "endDate", {
            /**
             * Date range end.
             * @return {?}
             */
            get: function () { return this._endDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._endDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "startAt", {
            /**
             * A date representing the period (month or year) to start the calendar in.
             * @return {?}
             */
            get: function () { return this._startAt; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._startAt = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "selected", {
            /**
             * The currently selected date.
             * @return {?}
             */
            get: function () { return this._selected; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._selected = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "minDate", {
            /**
             * The minimum selectable date.
             * @return {?}
             */
            get: function () { return this._minDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._minDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "maxDate", {
            /**
             * The maximum selectable date.
             * @return {?}
             */
            get: function () { return this._maxDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._maxDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "activeDate", {
            /**
             * The current active date. This determines which time period is shown and which date is
             * highlighted when using keyboard navigation.
             * @return {?}
             */
            get: function () { return this._clampedActiveDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._clampedActiveDate = this._dateAdapter.clampDate(value, this.minDate, this.maxDate);
                this.stateChanges.next();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendar.prototype, "currentView", {
            /**
             * Whether the calendar is in month view.
             * @return {?}
             */
            get: function () { return this._currentView; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._currentView = value;
                this._moveFocusOnNextTick = true;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SatCalendar.prototype.ngAfterContentInit = function () {
            this._calendarHeaderPortal = new portal.ComponentPortal(this.headerComponent || SatCalendarHeader);
            this._initStartEnd();
            // Assign to the private property since we don't want to move focus on init.
            this._currentView = this.startView;
        };
        /**
         * @return {?}
         */
        SatCalendar.prototype._initStartEnd = function () {
            var /** @type {?} */ _today = this._dateAdapter.today();
            var /** @type {?} */ _end = this.endDate || _today;
            var /** @type {?} */ _start = this.beginDate || _today;
            this.activeDate = ((!this.nextMonth)
                ?
                    _start
                :
                    (this._dateAdapter.getMonth(_end) === this._dateAdapter.getMonth(_start)
                        ?
                            this._dateAdapter.addCalendarMonths(_end, 1)
                        :
                            _end));
        };
        /**
         * @return {?}
         */
        SatCalendar.prototype.ngAfterViewChecked = function () {
            if (this._moveFocusOnNextTick) {
                this._moveFocusOnNextTick = false;
                this.focusActiveCell();
            }
        };
        /**
         * @return {?}
         */
        SatCalendar.prototype.ngOnDestroy = function () {
            this._intlChanges.unsubscribe();
            this.stateChanges.complete();
        };
        /**
         * @param {?} changes
         * @return {?}
         */
        SatCalendar.prototype.ngOnChanges = function (changes) {
            var /** @type {?} */ change = changes["minDate"] || changes["maxDate"] || changes["dateFilter"];
            if (change && !change.firstChange) {
                var /** @type {?} */ view = this._getCurrentViewComponent();
                if (view) {
                    view._init();
                }
            }
            var /** @type {?} */ changeDate = changes["beginDate"] || changes["endDate"];
            if (changeDate) {
                this._initStartEnd();
            }
            this.stateChanges.next();
        };
        /**
         * @return {?}
         */
        SatCalendar.prototype.focusActiveCell = function () {
            this._getCurrentViewComponent()._focusActiveCell();
        };
        /**
         * Handles year selection in the multiyear view.
         * @param {?} normalizedYear
         * @return {?}
         */
        SatCalendar.prototype._yearSelectedInMultiYearView = function (normalizedYear) {
            this.yearSelected.emit(normalizedYear);
        };
        /**
         * Handles month selection in the year view.
         * @param {?} normalizedMonth
         * @return {?}
         */
        SatCalendar.prototype._monthSelectedInYearView = function (normalizedMonth) {
            this.monthSelected.emit(normalizedMonth);
        };
        /**
         * @return {?}
         */
        SatCalendar.prototype._userSelected = function () {
            // this._userSelection.emit();
        };
        /**
         * Handles year/month selection in the multi-year/year views.
         * @param {?} date
         * @param {?} view
         * @return {?}
         */
        SatCalendar.prototype._goToDateInView = function (date, view) {
            this.activeDate = date;
            this.currentView = view;
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        SatCalendar.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        /**
         * Returns the component instance that corresponds to the current calendar view.
         * @return {?}
         */
        SatCalendar.prototype._getCurrentViewComponent = function () {
            return this.monthView || this.yearView || this.multiYearView;
        };
        return SatCalendar;
    }());
    SatCalendar.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'sat-calendar',
                    template: "<ng-template [cdkPortalOutlet]=\"_calendarHeaderPortal\"></ng-template>\n<div class=\"mat-calendar-content\" [ngSwitch]=\"currentView\" cdkMonitorSubtreeFocus tabindex=\"-1\">\n  <sat-month-view\n      *ngSwitchCase=\"'month'\"\n      [(activeDate)]=\"activeDate\"\n      [selected]=\"selected\"\n      [beginDate]=\"beginDate\"\n      [endDate]=\"endDate\"\n      [rangeMode]=\"rangeMode\"\n      [dateFilter]=\"dateFilter\"\n      [maxDate]=\"maxDate\"\n      [minDate]=\"minDate\"\n      (selectedChange)=\"dateSelected.emit($event)\"\n      (_userSelection)=\"_userSelected()\">\n  </sat-month-view>\n\n  <sat-year-view\n      *ngSwitchCase=\"'year'\"\n      [activeDate]=\"activeDate\"\n      [selected]=\"selected\"\n      [dateFilter]=\"dateFilter\"\n      [maxDate]=\"maxDate\"\n      [minDate]=\"minDate\"\n      (monthSelected)=\"_monthSelectedInYearView($event)\"\n      (selectedChange)=\"_goToDateInView($event, 'month')\">\n  </sat-year-view>\n\n  <sat-multi-year-view\n      *ngSwitchCase=\"'multi-year'\"\n      [activeDate]=\"activeDate\"\n      [selected]=\"selected\"\n      [dateFilter]=\"dateFilter\"\n      [maxDate]=\"maxDate\"\n      [minDate]=\"minDate\"\n      (yearSelected)=\"_yearSelectedInMultiYearView($event)\"\n      (selectedChange)=\"_goToDateInView($event, 'year')\">\n  </sat-multi-year-view>\n</div>\n",
                    styles: [".mat-calendar{display:block}.mat-calendar-header{padding:8px 8px 0}.mat-calendar-content{padding:0 8px 8px;outline:0}.mat-calendar-controls{display:flex;margin:5% calc(33% / 7 - 16px)}.mat-calendar-spacer{flex:1 1 auto}.mat-calendar-period-button{min-width:0}.mat-calendar-arrow{display:inline-block;width:0;height:0;border-left:5px solid transparent;border-right:5px solid transparent;border-top-width:5px;border-top-style:solid;margin:0 0 0 5px;vertical-align:middle}.mat-calendar-arrow.mat-calendar-invert{-webkit-transform:rotate(180deg);transform:rotate(180deg)}[dir=rtl] .mat-calendar-arrow{margin:0 5px 0 0}.mat-calendar-next-button,.mat-calendar-previous-button{position:relative}.mat-calendar-next-button::after,.mat-calendar-previous-button::after{top:0;left:0;right:0;bottom:0;position:absolute;content:'';margin:15.5px;border:0 solid currentColor;border-top-width:2px}[dir=rtl] .mat-calendar-next-button,[dir=rtl] .mat-calendar-previous-button{-webkit-transform:rotate(180deg);transform:rotate(180deg)}.mat-calendar-previous-button::after{border-left-width:2px;-webkit-transform:translateX(2px) rotate(-45deg);transform:translateX(2px) rotate(-45deg)}.mat-calendar-next-button::after{border-right-width:2px;-webkit-transform:translateX(-2px) rotate(45deg);transform:translateX(-2px) rotate(45deg)}.mat-calendar-table{border-spacing:0;border-collapse:collapse;width:100%}.mat-calendar-table-header th{text-align:center;padding:0 0 8px}.mat-calendar-table-header-divider{position:relative;height:1px}.mat-calendar-table-header-divider::after{content:'';position:absolute;top:0;left:-8px;right:-8px;height:1px}"],
                    host: {
                        'class': 'mat-calendar',
                    },
                    exportAs: 'matCalendar',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    SatCalendar.ctorParameters = function () {
        return [
            { type: matRangeDatepickerIntl, },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
            { type: core.ChangeDetectorRef, },
        ];
    };
    SatCalendar.propDecorators = {
        "beginDate": [{ type: core.Input },],
        "endDate": [{ type: core.Input },],
        "rangeMode": [{ type: core.Input },],
        "dateSelected": [{ type: core.Output },],
        "headerComponent": [{ type: core.Input },],
        "startAt": [{ type: core.Input },],
        "startView": [{ type: core.Input },],
        "selected": [{ type: core.Input },],
        "minDate": [{ type: core.Input },],
        "maxDate": [{ type: core.Input },],
        "dateFilter": [{ type: core.Input },],
        "selectedChange": [{ type: core.Output },],
        "yearSelected": [{ type: core.Output },],
        "monthSelected": [{ type: core.Output },],
        "_userSelection": [{ type: core.Output },],
        "monthView": [{ type: core.ViewChild, args: [SatMonthView,] },],
        "yearView": [{ type: core.ViewChild, args: [SatYearView,] },],
        "multiYearView": [{ type: core.ViewChild, args: [SatMultiYearView,] },],
        "nextMonth": [{ type: core.Input },],
        "activeDate": [{ type: core.Input },],
    };
    /**
     * Default header for SatCalendar
     * @template D
     */
    var SatCalendarHeader = (function () {
        /**
         * @param {?} _intl
         * @param {?} calendar
         * @param {?} _dateAdapter
         * @param {?} _dateFormats
         * @param {?} changeDetectorRef
         */
        function SatCalendarHeader(_intl, calendar, _dateAdapter, _dateFormats, changeDetectorRef) {
            this._intl = _intl;
            this.calendar = calendar;
            this._dateAdapter = _dateAdapter;
            this._dateFormats = _dateFormats;
            this.calendar.stateChanges.subscribe(function () { return changeDetectorRef.markForCheck(); });
        }
        Object.defineProperty(SatCalendarHeader.prototype, "periodButtonText", {
            /**
             * The label for the current calendar view.
             * @return {?}
             */
            get: function () {
                if (this.calendar.currentView == 'month') {
                    return this._dateAdapter
                        .format(this.calendar.activeDate, this._dateFormats.display.monthYearLabel)
                        .toLocaleUpperCase();
                }
                if (this.calendar.currentView == 'year') {
                    return this._dateAdapter.getYearName(this.calendar.activeDate);
                }
                var /** @type {?} */ activeYear = this._dateAdapter.getYear(this.calendar.activeDate);
                var /** @type {?} */ firstYearInView = this._dateAdapter.getYearName(this._dateAdapter.createDate(activeYear - activeYear % 24, 0, 1));
                var /** @type {?} */ lastYearInView = this._dateAdapter.getYearName(this._dateAdapter.createDate(activeYear + yearsPerPage - 1 - activeYear % 24, 0, 1));
                return firstYearInView + " \u2013 " + lastYearInView;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendarHeader.prototype, "periodButtonLabel", {
            /**
             * @return {?}
             */
            get: function () {
                return this.calendar.currentView == 'month' ?
                    this._intl.switchToMultiYearViewLabel : this._intl.switchToMonthViewLabel;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendarHeader.prototype, "prevButtonLabel", {
            /**
             * The label for the the previous button.
             * @return {?}
             */
            get: function () {
                return {
                    'month': this._intl.prevMonthLabel,
                    'year': this._intl.prevYearLabel,
                    'multi-year': this._intl.prevMultiYearLabel
                }[this.calendar.currentView];
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SatCalendarHeader.prototype, "nextButtonLabel", {
            /**
             * The label for the the next button.
             * @return {?}
             */
            get: function () {
                return {
                    'month': this._intl.nextMonthLabel,
                    'year': this._intl.nextYearLabel,
                    'multi-year': this._intl.nextMultiYearLabel
                }[this.calendar.currentView];
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Handles user clicks on the period label.
         * @return {?}
         */
        SatCalendarHeader.prototype.currentPeriodClicked = function () {
            this.calendar.currentView = this.calendar.currentView == 'month' ? 'multi-year' : 'month';
        };
        /**
         * Handles user clicks on the previous button.
         * @return {?}
         */
        SatCalendarHeader.prototype.previousClicked = function () {
            this.calendar.activeDate = this.calendar.currentView == 'month' ?
                this._dateAdapter.addCalendarMonths(this.calendar.activeDate, -1) :
                this._dateAdapter.addCalendarYears(this.calendar.activeDate, this.calendar.currentView == 'year' ? -1 : -yearsPerPage);
        };
        /**
         * Handles user clicks on the next button.
         * @return {?}
         */
        SatCalendarHeader.prototype.nextClicked = function () {
            this.calendar.activeDate = this.calendar.currentView == 'month' ?
                this._dateAdapter.addCalendarMonths(this.calendar.activeDate, 1) :
                this._dateAdapter.addCalendarYears(this.calendar.activeDate, this.calendar.currentView == 'year' ? 1 : yearsPerPage);
        };
        /**
         * Whether the previous period button is enabled.
         * @return {?}
         */
        SatCalendarHeader.prototype.previousEnabled = function () {
            if (!this.calendar.minDate) {
                return true;
            }
            return !this.calendar.minDate ||
                !this._isSameView(this.calendar.activeDate, this.calendar.minDate);
        };
        /**
         * Whether the next period button is enabled.
         * @return {?}
         */
        SatCalendarHeader.prototype.nextEnabled = function () {
            return !this.calendar.maxDate ||
                !this._isSameView(this.calendar.activeDate, this.calendar.maxDate);
        };
        /**
         * Whether the two dates represent the same view in the current view mode (month or year).
         * @param {?} date1
         * @param {?} date2
         * @return {?}
         */
        SatCalendarHeader.prototype._isSameView = function (date1, date2) {
            if (this.calendar.currentView == 'month') {
                return this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2) &&
                    this._dateAdapter.getMonth(date1) == this._dateAdapter.getMonth(date2);
            }
            if (this.calendar.currentView == 'year') {
                return this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2);
            }
            // Otherwise we are in 'multi-year' view.
            return Math.floor(this._dateAdapter.getYear(date1) / yearsPerPage) ==
                Math.floor(this._dateAdapter.getYear(date2) / yearsPerPage);
        };
        return SatCalendarHeader;
    }());
    SatCalendarHeader.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'sat-calendar-header',
                    template: "<div class=\"mat-calendar-header\">\n  <div class=\"mat-calendar-controls\">\n    <button mat-button type=\"button\" class=\"mat-calendar-period-button\"\n            (click)=\"currentPeriodClicked()\" [attr.aria-label]=\"periodButtonLabel\">\n      {{periodButtonText}}\n      <div class=\"mat-calendar-arrow\" [class.mat-calendar-invert]=\"calendar.currentView != 'month'\"></div>\n    </button>\n\n    <div class=\"mat-calendar-spacer\"></div>\n\n    <button mat-icon-button type=\"button\" class=\"mat-calendar-previous-button\"\n            [disabled]=\"!previousEnabled()\" (click)=\"previousClicked()\"\n            [attr.aria-label]=\"prevButtonLabel\">\n    </button>\n\n    <button mat-icon-button type=\"button\" class=\"mat-calendar-next-button\"\n            [disabled]=\"!nextEnabled()\" (click)=\"nextClicked()\"\n            [attr.aria-label]=\"nextButtonLabel\">\n    </button>\n  </div>\n</div>\n",
                    exportAs: 'matCalendarHeader',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    SatCalendarHeader.ctorParameters = function () {
        return [
            { type: matRangeDatepickerIntl, },
            { type: SatCalendar, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return SatCalendar; }),] },] },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
            { type: core.ChangeDetectorRef, },
        ];
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * A calendar that is used as part of the datepicker.
     * \@docs-private
     * @template D
     */
    var RangeCalendar = (function () {
        /**
         * @param {?} _intl
         * @param {?} _dateAdapter
         * @param {?} _dateFormats
         */
        function RangeCalendar(_intl, _dateAdapter, _dateFormats) {
            this._dateAdapter = _dateAdapter;
            this._dateFormats = _dateFormats;
            this.selectedChange = new core.EventEmitter();
            this.dateRangesChange = new core.EventEmitter();
            this.yearSelected = new core.EventEmitter();
            this.monthSelected = new core.EventEmitter();
            this._userSelection = new core.EventEmitter();
            /**
             * Whenever user already selected start of dates interval.
             */
            this._beginDateSelected = false;
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            if (!this._dateFormats) {
                throw createMissingDateImplError('MAT_DATE_FORMATS');
            }
        }
        /**
         * @param {?} date
         * @return {?}
         */
        RangeCalendar.prototype._dateSelected = function (date) {
            if (this.rangeMode) {
                if (!this._beginDateSelected) {
                    this._beginDateSelected = true;
                    this.beginDate = date;
                    this.endDate = date;
                }
                else {
                    this._beginDateSelected = false;
                    if (this._dateAdapter.compareDate(/** @type {?} */ (this.beginDate), date) <= 0) {
                        this.dateRangesChange.emit({ begin: /** @type {?} */ (this.beginDate), end: date });
                    }
                    else {
                        this.dateRangesChange.emit({ begin: date, end: /** @type {?} */ (this.beginDate) });
                    }
                }
            }
            else if (!this._dateAdapter.sameDate(date, this.selected)) {
                this.selectedChange.emit(date);
            }
        };
        /**
         * @param {?} quarter
         * @return {?}
         */
        RangeCalendar.prototype._getQuarterDates = function (quarter) {
            var /** @type {?} */ _year = this._dateAdapter.getYear(this._dateAdapter.today());
            var /** @type {?} */ begin = this._dateAdapter.createDate(_year, (quarter - 1) * 3, 1);
            var /** @type {?} */ end = (quarter === 4
                ?
                    this._dateAdapter.addCalendarDays(this._dateAdapter.createDate(_year + 1, 0, 1), -1)
                :
                    this._dateAdapter.addCalendarDays(this._dateAdapter.createDate(_year, quarter * 3, 1), -1));
            return ({ begin: begin, end: end });
        };
        /**
         * @param {?} quarter
         * @return {?}
         */
        RangeCalendar.prototype.selectQuater = function (quarter) {
            this.dateRangesChange.emit(this._getQuarterDates(quarter));
        };
        /**
         * @param {?} quarter
         * @return {?}
         */
        RangeCalendar.prototype.checkQuarterSelected = function (quarter) {
            var _a = this._getQuarterDates(quarter), begin = _a.begin, end = _a.end;
            return (this._dateAdapter.sameDate(begin, this.beginDate)
                &&
                    this._dateAdapter.sameDate(end, this.endDate));
        };
        /**
         * @return {?}
         */
        RangeCalendar.prototype.onCancel = function () {
            this._userSelection.emit();
        };
        /**
         * @return {?}
         */
        RangeCalendar.prototype.onApply = function () {
            this._userSelection.emit({ apply: true });
        };
        return RangeCalendar;
    }());
    RangeCalendar.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'range-calendar',
                    template: "<div class=\"from-date-sec\">\n  <sat-calendar cdkTrapFocus\n    [id]=\"id\"\n    [ngClass]=\"ngClass\"\n    [startAt]=\"startAt\"\n    [startView]=\"startView\"\n    [minDate]=\"minDate\"\n    [maxDate]=\"maxDate\"\n    [dateFilter]=\"dateFilter\"\n    [beginDate]=\"beginDate\"\n    [endDate]=\"endDate\"\n    [rangeMode]=\"rangeMode\"           \n    [selected]=\"selected\"\n    (selectedChange)=\"selectedChange.emit($event)\"\n    (dateSelected)=\"_dateSelected($event)\"\n    (yearSelected)=\"yearSelected.emit($event)\"\n    (monthSelected)=\"monthSelected.emit($event)\"\n    >\n  </sat-calendar>\n</div>\n<div class=\"end-date-sec\">\n  <sat-calendar cdkTrapFocus\n    [id]=\"id\"\n    [ngClass]=\"ngClass\"\n    [startAt]=\"startAt\"\n    [startView]=\"startView\"\n    [minDate]=\"minDate\"\n    [maxDate]=\"maxDate\"\n    [dateFilter]=\"dateFilter\"\n    [beginDate]=\"beginDate\"\n    [endDate]=\"endDate\"\n    [rangeMode]=\"rangeMode\"\n    [selected]=\"selected\"\n    (selectedChange)=\"selectedChange.emit($event)\"\n    (dateSelected)=\"_dateSelected($event)\"\n    (yearSelected)=\"yearSelected.emit($event)\"\n    (monthSelected)=\"monthSelected.emit($event)\"\n    [nextMonth]=\"true\"\n    >\n  </sat-calendar>\n</div>\n<div class=\"quick-select-sec\">\n  <span class=\"select-title\">Quick Select</span>\n  <ul>\n    <li><button mat-button (click)=\"selectQuater(1)\" [class.active]=\"checkQuarterSelected(1)\">Quarter 1</button></li>\n    <li><button mat-button (click)=\"selectQuater(2)\" [class.active]=\"checkQuarterSelected(2)\">Quarter 2</button></li>\n    <li><button mat-button (click)=\"selectQuater(3)\" [class.active]=\"checkQuarterSelected(3)\">Quarter 3</button></li>\n    <li><button mat-button (click)=\"selectQuater(4)\" [class.active]=\"checkQuarterSelected(4)\">Quarter 4</button></li>\n  </ul>\n</div>\n<div class=\"calendar-button-sec\">\n  <button mat-button color=\"primary\" (click)=\"onCancel()\">Cancel</button>\n  <button mat-raised-button color=\"primary\" (click)=\"onApply()\">Apply</button>\n</div>",
                    styles: [".mat-calendar{display:block}.mat-calendar-header{padding:8px 8px 0}.mat-calendar-content{padding:0 8px 8px;outline:0}.mat-calendar-controls{display:flex;margin:5% calc(33% / 7 - 16px)}.mat-calendar-spacer{flex:1 1 auto}.mat-calendar-period-button{min-width:0}.mat-calendar-arrow{display:inline-block;width:0;height:0;border-left:5px solid transparent;border-right:5px solid transparent;border-top-width:5px;border-top-style:solid;margin:0 0 0 5px;vertical-align:middle}.mat-calendar-arrow.mat-calendar-invert{-webkit-transform:rotate(180deg);transform:rotate(180deg)}[dir=rtl] .mat-calendar-arrow{margin:0 5px 0 0}.mat-calendar-next-button,.mat-calendar-previous-button{position:relative}.mat-calendar-next-button::after,.mat-calendar-previous-button::after{top:0;left:0;right:0;bottom:0;position:absolute;content:'';margin:15.5px;border:0 solid currentColor;border-top-width:2px}[dir=rtl] .mat-calendar-next-button,[dir=rtl] .mat-calendar-previous-button{-webkit-transform:rotate(180deg);transform:rotate(180deg)}.mat-calendar-previous-button::after{border-left-width:2px;-webkit-transform:translateX(2px) rotate(-45deg);transform:translateX(2px) rotate(-45deg)}.mat-calendar-next-button::after{border-right-width:2px;-webkit-transform:translateX(-2px) rotate(45deg);transform:translateX(-2px) rotate(45deg)}.mat-calendar-table{border-spacing:0;border-collapse:collapse;width:100%}.mat-calendar-table-header th{text-align:center;padding:0 0 8px}.mat-calendar-table-header-divider{position:relative;height:1px}.mat-calendar-table-header-divider::after{content:'';position:absolute;top:0;left:-8px;right:-8px;height:1px}"],
                    host: {
                        'class': 'range-calendar',
                    },
                    exportAs: 'matRangeCalendar',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    RangeCalendar.ctorParameters = function () {
        return [
            { type: matRangeDatepickerIntl, },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
        ];
    };
    RangeCalendar.propDecorators = {
        "id": [{ type: core.Input },],
        "ngClass": [{ type: core.Input },],
        "startAt": [{ type: core.Input },],
        "startView": [{ type: core.Input },],
        "minDate": [{ type: core.Input },],
        "maxDate": [{ type: core.Input },],
        "dateFilter": [{ type: core.Input },],
        "beginDate": [{ type: core.Input },],
        "endDate": [{ type: core.Input },],
        "rangeMode": [{ type: core.Input },],
        "selected": [{ type: core.Input },],
        "selectedChange": [{ type: core.Output },],
        "dateRangesChange": [{ type: core.Output },],
        "yearSelected": [{ type: core.Output },],
        "monthSelected": [{ type: core.Output },],
        "_userSelection": [{ type: core.Output },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * Animations used by the Material datepicker.
     */
    var /** @type {?} */ matDatepickerAnimations = {
        /** Transforms the height of the datepicker's calendar. */
        transformPanel: animations.trigger('transformPanel', [
            animations.state('void', animations.style({ opacity: 0, transform: 'scale(1, 0)', 'transform-origin': '0 0' })),
            animations.state('enter', animations.style({ opacity: 1, transform: 'scale(1, 1)', 'transform-origin': '0 0' })),
            animations.transition('void => enter', animations.animate('400ms cubic-bezier(0.25, 0.8, 0.25, 1)')),
            animations.transition('* => void', animations.animate('100ms linear', animations.style({ opacity: 0 }))),
        ]),
        /** Fades in the content of the calendar. */
        fadeInCalendar: animations.trigger('fadeInCalendar', [
            animations.state('void', animations.style({ opacity: 0 })),
            animations.state('enter', animations.style({ opacity: 1 })),
            animations.transition('void => *', animations.animate('400ms 100ms cubic-bezier(0.55, 0, 0.55, 0.2)')),
        ]),
    };
    /**
     * Used to generate a unique ID for each datepicker instance.
     */
    var /** @type {?} */ datepickerUid = 0;
    /**
     * Injection token that determines the scroll handling while the calendar is open.
     */
    var /** @type {?} */ MAT_DATEPICKER_SCROLL_STRATEGY = new core.InjectionToken('mat-range-datepicker-scroll-strategy', {
        providedIn: 'root',
        factory: MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY,
    });
    /**
     * \@docs-private
     * @return {?}
     */
    function MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY() {
        var /** @type {?} */ overlay$$1 = core.inject(overlay.Overlay);
        return function () { return overlay$$1.scrollStrategies.reposition(); };
    }
    /**
     * \@docs-private
     */
    var matRangeDatepickerContentBase = (function () {
        /**
         * @param {?} _elementRef
         */
        function matRangeDatepickerContentBase(_elementRef) {
            this._elementRef = _elementRef;
        }
        return matRangeDatepickerContentBase;
    }());
    var /** @type {?} */ _matRangeDatepickerContentMixinBase = core$1.mixinColor(matRangeDatepickerContentBase);
    /**
     * Component used as the content for the datepicker dialog and popup. We use this instead of using
     * SatCalendar directly as the content so we can control the initial focus. This also gives us a
     * place to put additional features of the popup that are not part of the calendar itself in the
     * future. (e.g. confirmation buttons).
     * \@docs-private
     * @template D
     */
    var matRangeDatepickerContent = (function (_super) {
        tslib.__extends(matRangeDatepickerContent, _super);
        /**
         * @param {?} elementRef
         * @param {?} _changeDetectorRef
         * @param {?} _ngZone
         */
        function matRangeDatepickerContent(elementRef, _changeDetectorRef, _ngZone) {
            var _this = _super.call(this, elementRef) || this;
            _this._changeDetectorRef = _changeDetectorRef;
            _this._ngZone = _ngZone;
            return _this;
        }
        /**
         * @return {?}
         */
        matRangeDatepickerContent.prototype.ngOnInit = function () {
            var _this = this;
            if (!this.datepicker._popupRef || this._positionChange) {
                return;
            }
            var /** @type {?} */ positionStrategy = (((this.datepicker._popupRef.getConfig().positionStrategy)));
            this._positionChange = positionStrategy.positionChanges.subscribe(function (change) {
                var /** @type {?} */ isAbove = change.connectionPair.overlayY === 'bottom';
                if (isAbove !== _this._isAbove) {
                    _this._ngZone.run(function () {
                        _this._isAbove = isAbove;
                        _this._changeDetectorRef.markForCheck();
                    });
                }
            });
        };
        /**
         * @return {?}
         */
        matRangeDatepickerContent.prototype.ngAfterViewInit = function () {
            // this._calendar.focusActiveCell();
        };
        /**
         * @return {?}
         */
        matRangeDatepickerContent.prototype.ngOnDestroy = function () {
            if (this._positionChange) {
                this._positionChange.unsubscribe();
                this._positionChange = null;
            }
        };
        return matRangeDatepickerContent;
    }(_matRangeDatepickerContentMixinBase));
    matRangeDatepickerContent.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'mat-range-datepicker-content',
                    template: "<range-calendar cdkTrapFocus\n              [id]=\"datepicker.id\"\n              [ngClass]=\"datepicker.panelClass\"\n              [startAt]=\"datepicker.startAt\"\n              [startView]=\"datepicker.startView\"\n              [minDate]=\"datepicker._minDate\"\n              [maxDate]=\"datepicker._maxDate\"\n              [dateFilter]=\"datepicker._dateFilter\"\n              [beginDate]=\"datepicker._beginDate\"\n              [endDate]=\"datepicker._endDate\"\n              [rangeMode]=\"datepicker.rangeMode\"\n              [selected]=\"datepicker._selected\"\n              (selectedChange)=\"datepicker._select($event)\"\n              (dateRangesChange)=\"datepicker._selectRange($event)\"\n              (yearSelected)=\"datepicker._selectYear($event)\"\n              (monthSelected)=\"datepicker._selectMonth($event)\"\n              (_userSelection)=\"datepicker.close($event)\">\n</range-calendar>\n",
                    styles: [".mat-datepicker-content{box-shadow:0 5px 5px -3px rgba(0,0,0,.2),0 8px 10px 1px rgba(0,0,0,.14),0 3px 14px 2px rgba(0,0,0,.12);display:block;width:593px;height:493px}.end-date-sec,.from-date-sec{display:inline-block;width:50%;height:360px;box-sizing:border-box}.from-date-sec{border-right:.5px solid #e9e9e9}.end-date-sec{border-left:0 solid #e9e9e9}.quick-select-sec{border-top:1px solid #e9e9e9;border-bottom:1px solid #e9e9e9;height:50px;line-height:50px;padding:0 20px;font-size:14px;font-weight:200}.quick-select-sec .select-title{color:#9f9f9f}.quick-select-sec ul{display:inline-block;list-style-type:none;margin:0;color:#696969}.quick-select-sec ul li{display:inline-block;margin-left:10px}.quick-select-sec ul li .mat-button{font-size:13px;font-weight:400}.calendar-button-sec{padding:0 20px;text-align:right;line-height:80px}.calendar-button-sec button{margin-left:10px}.mat-datepicker-content-touch{box-shadow:0 0 0 0 rgba(0,0,0,.2),0 0 0 0 rgba(0,0,0,.14),0 0 0 0 rgba(0,0,0,.12);display:block;max-height:80vh;overflow:auto;margin:-24px}.mat-datepicker-content-touch .mat-calendar{min-width:250px;min-height:312px;max-width:750px;max-height:788px}@media all and (orientation:landscape){.mat-datepicker-content-touch .mat-calendar{width:64vh;height:80vh}}@media all and (orientation:portrait){.mat-datepicker-content-touch .mat-calendar{width:80vw;height:100vw}}.cdk-keyboard-focused .mat-calendar-body-active>.mat-calendar-body-cell-content:not(.mat-calendar-body-selected):not(.mat-calendar-body-semi-selected),.cdk-program-focused .mat-calendar-body-active>.mat-calendar-body-cell-content:not(.mat-calendar-body-selected):not(.mat-calendar-body-semi-selected),:not(.mat-calendar-body-disabled):hover>.mat-calendar-body-cell-content:not(.mat-calendar-body-selected):not(.mat-calendar-body-semi-selected){background-color:rgba(0,0,0,.04)}.cdk-keyboard-focused .mat-calendar-body-active>.mat-calendar-body-semi-selected,.cdk-program-focused .mat-calendar-body-active>.mat-calendar-body-semi-selected,:not(.mat-calendar-body-disabled):hover>.mat-calendar-body-semi-selected{background-color:#3f51b5;color:#fff}.mat-calendar-body-begin-range:not(.mat-calendar-body-end-range){border-radius:100% 0 0 100%;background-color:#e8eaf6}.mat-calendar-body-end-range:not(.mat-calendar-body-begin-range){border-radius:0 100% 100% 0;background-color:#e8eaf6}.mat-calendar-body>tr .mat-calendar-cell-semi-selected~.mat-calendar-cell-semi-selected{border-radius:0}.mat-calendar-cell-semi-selected{background-color:#f0e7ff}.mat-button.active .mat-button-focus-overlay{opacity:1}"],
                    host: {
                        'class': 'mat-datepicker-content',
                        '[@transformPanel]': '"enter"',
                        '[class.mat-datepicker-content-touch]': 'datepicker.touchUi',
                        '[class.mat-datepicker-content-above]': '_isAbove',
                    },
                    animations: [
                        matDatepickerAnimations.transformPanel,
                        matDatepickerAnimations.fadeInCalendar,
                    ],
                    exportAs: 'matDatepickerContent',
                    encapsulation: core.ViewEncapsulation.None,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    inputs: ['color'],
                },] },
    ];
    /** @nocollapse */
    matRangeDatepickerContent.ctorParameters = function () {
        return [
            { type: core.ElementRef, },
            { type: core.ChangeDetectorRef, },
            { type: core.NgZone, },
        ];
    };
    matRangeDatepickerContent.propDecorators = {
        "_calendar": [{ type: core.ViewChild, args: [RangeCalendar,] },],
    };
    /**
     * Component responsible for managing the datepicker popup/dialog.
     * @template D
     */
    var matRangeDatepicker = (function () {
        /**
         * @param {?} _dialog
         * @param {?} _overlay
         * @param {?} _ngZone
         * @param {?} _viewContainerRef
         * @param {?} _scrollStrategy
         * @param {?} _dateAdapter
         * @param {?} _dir
         * @param {?} _document
         */
        function matRangeDatepicker(_dialog, _overlay, _ngZone, _viewContainerRef, _scrollStrategy, _dateAdapter, _dir, _document) {
            this._dialog = _dialog;
            this._overlay = _overlay;
            this._ngZone = _ngZone;
            this._viewContainerRef = _viewContainerRef;
            this._scrollStrategy = _scrollStrategy;
            this._dateAdapter = _dateAdapter;
            this._dir = _dir;
            this._document = _document;
            /**
             * The view that the calendar should start in.
             */
            this.startView = 'month';
            this._touchUi = false;
            /**
             * Emits selected year in multiyear view.
             * This doesn't imply a change on the selected date.
             */
            this.yearSelected = new core.EventEmitter();
            /**
             * Emits selected month in year view.
             * This doesn't imply a change on the selected date.
             */
            this.monthSelected = new core.EventEmitter();
            /**
             * Emits when the datepicker has been opened.
             */
            this.openedStream = new core.EventEmitter();
            /**
             * Emits when the datepicker has been closed.
             */
            this.closedStream = new core.EventEmitter();
            this._opened = false;
            /**
             * The id for the datepicker calendar.
             */
            this.id = "mat-range-datepicker-" + datepickerUid++;
            this._validSelected = null;
            /**
             * The element that was focused before the datepicker was opened.
             */
            this._focusedElementBeforeOpen = null;
            /**
             * Subscription to value changes in the associated input element.
             */
            this._inputSubscription = rxjs.Subscription.EMPTY;
            /**
             * Emits when the datepicker is disabled.
             */
            this._disabledChange = new rxjs.Subject();
            /**
             * Emits new selected date when selected date changes.
             */
            this._selectedChanged = new rxjs.Subject();
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
        }
        Object.defineProperty(matRangeDatepicker.prototype, "rangeMode", {
            /**
             * Whenever datepicker is for selecting range of dates.
             * @return {?}
             */
            get: function () {
                return this._rangeMode;
            },
            /**
             * @param {?} mode
             * @return {?}
             */
            set: function (mode) {
                this._rangeMode = mode;
                if (this.rangeMode) {
                    this._validSelected = null;
                }
                else {
                    this._beginDate = this._endDate = null;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "beginDate", {
            /**
             * Start of dates interval.
             * @return {?}
             */
            get: function () { return this._beginDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._validSelected = null;
                this._beginDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "endDate", {
            /**
             * End of dates interval.
             * @return {?}
             */
            get: function () { return this._endDate; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._validSelected = null;
                this._endDate = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "startAt", {
            /**
             * The date to open the calendar to initially.
             * @return {?}
             */
            get: function () {
                // If an explicit startAt is set we start there, otherwise we start at whatever the currently
                // selected value is.
                if (this.rangeMode) {
                    return this._startAt || (this._datepickerInput && this._datepickerInput.value ?
                        ((this._datepickerInput.value)).begin : null);
                }
                return this._startAt || (this._datepickerInput ? /** @type {?} */ (this._datepickerInput.value) : null);
            },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._startAt = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "color", {
            /**
             * Color palette to use on the datepicker's calendar.
             * @return {?}
             */
            get: function () {
                return this._color ||
                    (this._datepickerInput ? this._datepickerInput._getThemePalette() : undefined);
            },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._color = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "touchUi", {
            /**
             * Whether the calendar UI is in touch mode. In touch mode the calendar opens in a dialog rather
             * than a popup and elements have more padding to allow for bigger touch targets.
             * @return {?}
             */
            get: function () { return this._touchUi; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._touchUi = coercion.coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "disabled", {
            /**
             * Whether the datepicker pop-up should be disabled.
             * @return {?}
             */
            get: function () {
                return this._disabled === undefined && this._datepickerInput ?
                    this._datepickerInput.disabled : !!this._disabled;
            },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                var /** @type {?} */ newValue = coercion.coerceBooleanProperty(value);
                if (newValue !== this._disabled) {
                    this._disabled = newValue;
                    this._disabledChange.next(newValue);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "opened", {
            /**
             * Whether the calendar is open.
             * @return {?}
             */
            get: function () { return this._opened; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) { value ? this.open() : this.close(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "_selected", {
            /**
             * The currently selected date.
             * @return {?}
             */
            get: function () { return this._validSelected; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) { this._validSelected = value; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "_minDate", {
            /**
             * The minimum selectable date.
             * @return {?}
             */
            get: function () {
                return this._datepickerInput && this._datepickerInput.min;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "_maxDate", {
            /**
             * The maximum selectable date.
             * @return {?}
             */
            get: function () {
                return this._datepickerInput && this._datepickerInput.max;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepicker.prototype, "_dateFilter", {
            /**
             * @return {?}
             */
            get: function () {
                return this._datepickerInput && this._datepickerInput._dateFilter;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        matRangeDatepicker.prototype.ngAfterContentInit = function () {
            this._initBeginDate = this.beginDate;
            this._initEndDate = this.endDate;
        };
        /**
         * @return {?}
         */
        matRangeDatepicker.prototype.ngOnDestroy = function () {
            this.close();
            this._inputSubscription.unsubscribe();
            this._disabledChange.complete();
            if (this._popupRef) {
                this._popupRef.dispose();
                this._popupComponentRef = null;
            }
        };
        /**
         * Selects the given date
         * @param {?} date
         * @return {?}
         */
        matRangeDatepicker.prototype._select = function (date) {
            var /** @type {?} */ oldValue = this._selected;
            this._selected = date;
            if (!this._dateAdapter.sameDate(oldValue, this._selected)) {
                this._selectedChanged.next(date);
            }
        };
        /**
         * Selects the given date range
         * @param {?} dates
         * @return {?}
         */
        matRangeDatepicker.prototype._selectRange = function (dates) {
            if (!this._dateAdapter.sameDate(dates.begin, this.beginDate) ||
                !this._dateAdapter.sameDate(dates.end, this.endDate)) {
                this._selectedChanged.next(dates);
            }
            this._beginDate = dates.begin;
            this._endDate = dates.end;
        };
        /**
         * Emits the selected year in multiyear view
         * @param {?} normalizedYear
         * @return {?}
         */
        matRangeDatepicker.prototype._selectYear = function (normalizedYear) {
            this.yearSelected.emit(normalizedYear);
        };
        /**
         * Emits selected month in year view
         * @param {?} normalizedMonth
         * @return {?}
         */
        matRangeDatepicker.prototype._selectMonth = function (normalizedMonth) {
            this.monthSelected.emit(normalizedMonth);
        };
        /**
         * Register an input with this datepicker.
         * @param {?} input The datepicker input to register with this datepicker.
         * @return {?}
         */
        matRangeDatepicker.prototype._registerInput = function (input$$1) {
            var _this = this;
            if (this._datepickerInput) {
                throw Error('A matRangeDatepicker can only be associated with a single input.');
            }
            this._datepickerInput = input$$1;
            this._inputSubscription =
                this._datepickerInput._valueChange
                    .subscribe(function (value) {
                    if (value === null) {
                        _this.beginDate = _this.endDate = _this._selected = null;
                        return;
                    }
                    if (_this.rangeMode) {
                        value = /** @type {?} */ (value);
                        if (value.begin && value.end &&
                            _this._dateAdapter.compareDate(value.begin, value.end) <= 0) {
                            _this.beginDate = value.begin;
                            _this.endDate = value.end;
                        }
                        else {
                            _this.beginDate = _this.endDate = null;
                        }
                    }
                    else {
                        _this._selected = /** @type {?} */ (value);
                    }
                });
        };
        /**
         * Open the calendar.
         * @return {?}
         */
        matRangeDatepicker.prototype.open = function () {
            this._initBeginDate = this.beginDate;
            this._initEndDate = this.endDate;
            if (this._opened || this.disabled) {
                return;
            }
            if (!this._datepickerInput) {
                throw Error('Attempted to open an matRangeDatepicker with no associated input.');
            }
            if (this._document) {
                this._focusedElementBeforeOpen = this._document.activeElement;
            }
            this.touchUi ? this._openAsDialog() : this._openAsPopup();
            this._opened = true;
            this.openedStream.emit();
        };
        /**
         * Close the calendar.
         * @param {?=} options
         * @return {?}
         */
        matRangeDatepicker.prototype.close = function (options) {
            var _this = this;
            if (!this._opened) {
                return;
            }
            if (this._popupRef && this._popupRef.hasAttached()) {
                this._popupRef.detach();
            }
            if (this._dialogRef) {
                this._dialogRef.close();
                this._dialogRef = null;
            }
            if (this._calendarPortal && this._calendarPortal.isAttached) {
                this._calendarPortal.detach();
            }
            //restore if not applied
            if (!options || !options.apply) {
                this.beginDate = this._initBeginDate;
                this.endDate = this._initEndDate;
                this._selectedChanged.next({
                    begin: this.beginDate,
                    end: this.endDate,
                });
            }
            var /** @type {?} */ completeClose = function () {
                // The `_opened` could've been reset already if
                // we got two events in quick succession.
                if (_this._opened) {
                    _this._opened = false;
                    _this.closedStream.emit();
                    _this._focusedElementBeforeOpen = null;
                }
            };
            if (this._focusedElementBeforeOpen &&
                typeof this._focusedElementBeforeOpen.focus === 'function') {
                // Because IE moves focus asynchronously, we can't count on it being restored before we've
                // marked the datepicker as closed. If the event fires out of sequence and the element that
                // we're refocusing opens the datepicker on focus, the user could be stuck with not being
                // able to close the calendar at all. We work around it by making the logic, that marks
                // the datepicker as closed, async as well.
                this._focusedElementBeforeOpen.focus();
                setTimeout(completeClose);
            }
            else {
                completeClose();
            }
        };
        /**
         * Open the calendar as a dialog.
         * @return {?}
         */
        matRangeDatepicker.prototype._openAsDialog = function () {
            var _this = this;
            this._dialogRef = this._dialog.open(matRangeDatepickerContent, {
                direction: this._getDirection(),
                viewContainerRef: this._viewContainerRef,
                panelClass: 'mat-datepicker-dialog',
            });
            this._dialogRef.afterClosed().subscribe(function () { return _this.close(); });
            this._dialogRef.componentInstance.datepicker = this;
            this._setColor();
        };
        /**
         * Open the calendar as a popup.
         * @return {?}
         */
        matRangeDatepicker.prototype._openAsPopup = function () {
            var _this = this;
            if (!this._calendarPortal) {
                this._calendarPortal = new portal.ComponentPortal(matRangeDatepickerContent, this._viewContainerRef);
            }
            if (!this._popupRef) {
                this._createPopup();
            }
            if (!this._popupRef.hasAttached()) {
                this._popupRef.setDirection(this._getDirection());
                this._popupComponentRef = this._popupRef.attach(this._calendarPortal);
                this._popupComponentRef.instance.datepicker = this;
                this._setColor();
                // Update the position once the calendar has rendered.
                this._ngZone.onStable.asObservable().pipe(operators.take(1)).subscribe(function () {
                    _this._popupRef.updatePosition();
                });
            }
        };
        /**
         * Create the popup.
         * @return {?}
         */
        matRangeDatepicker.prototype._createPopup = function () {
            var _this = this;
            var /** @type {?} */ overlayConfig = new overlay.OverlayConfig({
                positionStrategy: this._createPopupPositionStrategy(),
                hasBackdrop: true,
                backdropClass: 'mat-overlay-transparent-backdrop',
                direction: this._getDirection(),
                scrollStrategy: this._scrollStrategy(),
                panelClass: 'mat-datepicker-popup',
            });
            this._popupRef = this._overlay.create(overlayConfig);
            rxjs.merge(this._popupRef.backdropClick(), 
            // this._popupRef.detachments(),
            this._popupRef.keydownEvents().pipe(operators.filter(function (event) {
                // Closing on alt + up is only valid when there's an input associated with the datepicker.
                return event.keyCode === keycodes.ESCAPE ||
                    (_this._datepickerInput && event.altKey && event.keyCode === keycodes.UP_ARROW);
            }))).subscribe(function () { return _this.close(); });
        };
        /**
         * Create the popup PositionStrategy.
         * @return {?}
         */
        matRangeDatepicker.prototype._createPopupPositionStrategy = function () {
            return this._overlay.position()
                .flexibleConnectedTo(this._datepickerInput.getPopupConnectionElementRef())
                .withFlexibleDimensions(false)
                .withViewportMargin(8)
                .withPush(false)
                .withPositions([
                {
                    originX: 'start',
                    originY: 'bottom',
                    overlayX: 'start',
                    overlayY: 'top'
                },
                {
                    originX: 'start',
                    originY: 'top',
                    overlayX: 'start',
                    overlayY: 'bottom'
                },
                {
                    originX: 'end',
                    originY: 'bottom',
                    overlayX: 'end',
                    overlayY: 'top'
                },
                {
                    originX: 'end',
                    originY: 'top',
                    overlayX: 'end',
                    overlayY: 'bottom'
                }
            ]);
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        matRangeDatepicker.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        /**
         * Passes the current theme color along to the calendar overlay.
         * @return {?}
         */
        matRangeDatepicker.prototype._setColor = function () {
            var /** @type {?} */ color = this.color;
            if (this._popupComponentRef) {
                this._popupComponentRef.instance.color = color;
            }
            if (this._dialogRef) {
                this._dialogRef.componentInstance.color = color;
            }
        };
        /**
         * Returns the layout direction of the datepicker.
         * @return {?}
         */
        matRangeDatepicker.prototype._getDirection = function () {
            return this._dir ? this._dir.value : 'ltr';
        };
        return matRangeDatepicker;
    }());
    matRangeDatepicker.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'mat-range-datepicker',
                    template: '',
                    exportAs: 'matDatepicker',
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    encapsulation: core.ViewEncapsulation.None,
                },] },
    ];
    /** @nocollapse */
    matRangeDatepicker.ctorParameters = function () {
        return [
            { type: dialog.MatDialog, },
            { type: overlay.Overlay, },
            { type: core.NgZone, },
            { type: core.ViewContainerRef, },
            { type: undefined, decorators: [{ type: core.Inject, args: [MAT_DATEPICKER_SCROLL_STRATEGY,] },] },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: bidi.Directionality, decorators: [{ type: core.Optional },] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [common.DOCUMENT,] },] },
        ];
    };
    matRangeDatepicker.propDecorators = {
        "rangeMode": [{ type: core.Input },],
        "beginDate": [{ type: core.Input },],
        "endDate": [{ type: core.Input },],
        "calendarHeaderComponent": [{ type: core.Input },],
        "startAt": [{ type: core.Input },],
        "startView": [{ type: core.Input },],
        "color": [{ type: core.Input },],
        "touchUi": [{ type: core.Input },],
        "disabled": [{ type: core.Input },],
        "yearSelected": [{ type: core.Output },],
        "monthSelected": [{ type: core.Output },],
        "panelClass": [{ type: core.Input },],
        "openedStream": [{ type: core.Output, args: ['opened',] },],
        "closedStream": [{ type: core.Output, args: ['closed',] },],
        "opened": [{ type: core.Input },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ MAT_DATEPICKER_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return matRangeDatepickerInput; }),
        multi: true
    };
    var /** @type {?} */ MAT_DATEPICKER_VALIDATORS = {
        provide: forms.NG_VALIDATORS,
        useExisting: core.forwardRef(function () { return matRangeDatepickerInput; }),
        multi: true
    };
    /**
     * An event used for datepicker input and change events. We don't always have access to a native
     * input or change event because the event may have been triggered by the user clicking on the
     * calendar popup. For consistency, we always use matRangeDatepickerInputEvent instead.
     * @template D
     */
    var matRangeDatepickerInputEvent = (function () {
        /**
         * @param {?} target
         * @param {?} targetElement
         */
        function matRangeDatepickerInputEvent(target, targetElement) {
            this.target = target;
            this.targetElement = targetElement;
            this.value = this.target.value;
        }
        return matRangeDatepickerInputEvent;
    }());
    /**
     * Directive used to connect an input to a matRangeDatepicker.
     * @template D
     */
    var matRangeDatepickerInput = (function () {
        /**
         * @param {?} _elementRef
         * @param {?} _dateAdapter
         * @param {?} _dateFormats
         * @param {?} _formField
         */
        function matRangeDatepickerInput(_elementRef, _dateAdapter, _dateFormats, _formField) {
            var _this = this;
            this._elementRef = _elementRef;
            this._dateAdapter = _dateAdapter;
            this._dateFormats = _dateFormats;
            this._formField = _formField;
            /**
             * Emits when a `change` event is fired on this `<input>`.
             */
            this.dateChange = new core.EventEmitter();
            /**
             * Emits when an `input` event is fired on this `<input>`.
             */
            this.dateInput = new core.EventEmitter();
            /**
             * Emits when the value changes (either due to user input or programmatic change).
             */
            this._valueChange = new core.EventEmitter();
            /**
             * Emits when the disabled state has changed
             */
            this._disabledChange = new core.EventEmitter();
            this._onTouched = function () { };
            this._cvaOnChange = function () { };
            this._validatorOnChange = function () { };
            this._datepickerSubscription = rxjs.Subscription.EMPTY;
            this._localeSubscription = rxjs.Subscription.EMPTY;
            /**
             * The form control validator for whether the input parses.
             */
            this._parseValidator = function () {
                return _this._lastValueValid ?
                    null : { 'matDatepickerParse': { 'text': _this._elementRef.nativeElement.value } };
            };
            /**
             * The form control validator for the min date.
             */
            this._minValidator = function (control) {
                if (_this._datepicker.rangeMode && control.value) {
                    var /** @type {?} */ beginDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.begin));
                    var /** @type {?} */ endDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.end));
                    if (_this.min) {
                        if (beginDate && _this._dateAdapter.compareDate(_this.min, beginDate) > 0) {
                            return { 'matDatepickerMin': { 'min': _this.min, 'actual': beginDate } };
                        }
                        if (endDate && _this._dateAdapter.compareDate(_this.min, endDate) > 0) {
                            return { 'matDatepickerMin': { 'min': _this.min, 'actual': endDate } };
                        }
                    }
                    return null;
                }
                var /** @type {?} */ controlValue = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value));
                return (!_this.min || !controlValue ||
                    _this._dateAdapter.compareDate(_this.min, controlValue) <= 0) ?
                    null : { 'matDatepickerMin': { 'min': _this.min, 'actual': controlValue } };
            };
            /**
             * The form control validator for the max date.
             */
            this._maxValidator = function (control) {
                if (_this._datepicker.rangeMode && control.value) {
                    var /** @type {?} */ beginDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.begin));
                    var /** @type {?} */ endDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.end));
                    if (_this.max) {
                        if (beginDate && _this._dateAdapter.compareDate(_this.max, beginDate) < 0) {
                            return { 'matDatepickerMax': { 'max': _this.max, 'actual': beginDate } };
                        }
                        if (endDate && _this._dateAdapter.compareDate(_this.max, endDate) < 0) {
                            return { 'matDatepickerMax': { 'max': _this.max, 'actual': endDate } };
                        }
                    }
                    return null;
                }
                var /** @type {?} */ controlValue = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value));
                return (!_this.max || !controlValue ||
                    _this._dateAdapter.compareDate(_this.max, controlValue) >= 0) ?
                    null : { 'matDatepickerMax': { 'max': _this.max, 'actual': controlValue } };
            };
            /**
             * The form control validator for the date filter.
             */
            this._filterValidator = function (control) {
                if (_this._datepicker.rangeMode && control.value) {
                    var /** @type {?} */ beginDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.begin));
                    var /** @type {?} */ endDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.end));
                    return !_this._dateFilter || !beginDate && !endDate ||
                        _this._dateFilter(beginDate) && _this._dateFilter(endDate) ?
                        null : { 'matDatepickerFilter': true };
                }
                var /** @type {?} */ controlValue = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value));
                return !_this._dateFilter || !controlValue || _this._dateFilter(controlValue) ?
                    null : { 'matDatepickerFilter': true };
            };
            /**
             * The form control validator for the date filter.
             */
            this._rangeValidator = function (control) {
                if (_this._datepicker.rangeMode && control.value) {
                    var /** @type {?} */ beginDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.begin));
                    var /** @type {?} */ endDate = _this._getValidDateOrNull(_this._dateAdapter.deserialize(control.value.end));
                    return !beginDate || !endDate || _this._dateAdapter.compareDate(beginDate, endDate) <= 0 ?
                        null : { 'matDatepickerRange': true };
                }
                return null;
            };
            /**
             * The combined form control validator for this input.
             */
            this._validator = forms.Validators.compose([this._parseValidator, this._minValidator, this._maxValidator,
                this._filterValidator, this._rangeValidator]);
            /**
             * Whether the last value set on the input was valid.
             */
            this._lastValueValid = false;
            if (!this._dateAdapter) {
                throw createMissingDateImplError('DateAdapter');
            }
            if (!this._dateFormats) {
                throw createMissingDateImplError('MAT_DATE_FORMATS');
            }
            // Update the displayed date when the locale changes.
            this._localeSubscription = _dateAdapter.localeChanges.subscribe(function () {
                _this.value = _this.value;
            });
        }
        Object.defineProperty(matRangeDatepickerInput.prototype, "matRangeDatepicker", {
            /**
             * The datepicker that this input is associated with.
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this.registerDatepicker(value);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} value
         * @return {?}
         */
        matRangeDatepickerInput.prototype.registerDatepicker = function (value) {
            if (value) {
                this._datepicker = value;
                this._datepicker._registerInput(this);
            }
        };
        Object.defineProperty(matRangeDatepickerInput.prototype, "matDatepickerFilter", {
            /**
             * Function that can be used to filter out dates within the datepicker.
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._dateFilter = value;
                this._validatorOnChange();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepickerInput.prototype, "value", {
            /**
             * The value of the input.
             * @return {?}
             */
            get: function () {
                return this._value;
            },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                if (value && value.hasOwnProperty('begin') && value.hasOwnProperty('end')) {
                    /**
                     * Range mode
                     */
                    var /** @type {?} */ rangeValue = (value);
                    rangeValue.begin = this._dateAdapter.deserialize(rangeValue.begin);
                    rangeValue.end = this._dateAdapter.deserialize(rangeValue.end);
                    this._lastValueValid = !rangeValue.begin || !rangeValue.end ||
                        this._dateAdapter.isValid(rangeValue.begin) && this._dateAdapter.isValid(rangeValue.end);
                    rangeValue.begin = this._getValidDateOrNull(rangeValue.begin);
                    rangeValue.end = this._getValidDateOrNull(rangeValue.end);
                    var /** @type {?} */ oldDate = (this.value);
                    this._elementRef.nativeElement.value =
                        rangeValue && rangeValue.begin && rangeValue.end
                            ? this._dateAdapter.format(rangeValue.begin, this._dateFormats.display.dateInput) +
                                ' - ' +
                                this._dateAdapter.format(rangeValue.end, this._dateFormats.display.dateInput)
                            : '';
                    if (oldDate == null && rangeValue != null || oldDate != null && rangeValue == null ||
                        !this._dateAdapter.sameDate(((oldDate)).begin, rangeValue.begin) ||
                        !this._dateAdapter.sameDate(((oldDate)).end, rangeValue.end)) {
                        if (rangeValue.end && rangeValue.begin &&
                            this._dateAdapter
                                .compareDate(rangeValue.begin, rangeValue.end) > 0) {
                            // if begin > end
                            value = null;
                        }
                        this._value = value;
                        this._valueChange.emit(value);
                    }
                }
                else {
                    /** Not range mode */
                    value = this._dateAdapter.deserialize(value);
                    this._lastValueValid = !value || this._dateAdapter.isValid(value);
                    value = this._getValidDateOrNull(value);
                    var /** @type {?} */ oldDate = this.value;
                    this._value = value;
                    this._elementRef.nativeElement.value =
                        value ? this._dateAdapter.format(value, this._dateFormats.display.dateInput) : '';
                    if (!this._dateAdapter.sameDate(/** @type {?} */ (oldDate), value)) {
                        this._valueChange.emit(value);
                    }
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepickerInput.prototype, "min", {
            /**
             * The minimum valid date.
             * @return {?}
             */
            get: function () { return this._min; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._min = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this._validatorOnChange();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepickerInput.prototype, "max", {
            /**
             * The maximum valid date.
             * @return {?}
             */
            get: function () { return this._max; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._max = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
                this._validatorOnChange();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(matRangeDatepickerInput.prototype, "disabled", {
            /**
             * Whether the datepicker-input is disabled.
             * @return {?}
             */
            get: function () { return !!this._disabled; },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                var /** @type {?} */ newValue = coercion.coerceBooleanProperty(value);
                var /** @type {?} */ element = this._elementRef.nativeElement;
                if (this._disabled !== newValue) {
                    this._disabled = newValue;
                    this._disabledChange.emit(newValue);
                }
                // We need to null check the `blur` method, because it's undefined during SSR.
                if (newValue && element.blur) {
                    // Normally, native input elements automatically blur if they turn disabled. This behavior
                    // is problematic, because it would mean that it triggers another change detection cycle,
                    // which then causes a changed after checked error if the input element was focused before.
                    element.blur();
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        matRangeDatepickerInput.prototype.ngAfterContentInit = function () {
            var _this = this;
            if (this._datepicker) {
                this._datepickerSubscription =
                    this._datepicker._selectedChanged.subscribe(function (selected) {
                        _this.value = selected;
                        _this._cvaOnChange(selected);
                        _this._onTouched();
                        _this.dateInput.emit(new matRangeDatepickerInputEvent(_this, _this._elementRef.nativeElement));
                        _this.dateChange.emit(new matRangeDatepickerInputEvent(_this, _this._elementRef.nativeElement));
                    });
            }
        };
        /**
         * @return {?}
         */
        matRangeDatepickerInput.prototype.ngOnDestroy = function () {
            this._datepickerSubscription.unsubscribe();
            this._localeSubscription.unsubscribe();
            this._valueChange.complete();
            this._disabledChange.complete();
        };
        /**
         * \@docs-private
         * @param {?} fn
         * @return {?}
         */
        matRangeDatepickerInput.prototype.registerOnValidatorChange = function (fn) {
            this._validatorOnChange = fn;
        };
        /**
         * \@docs-private
         * @param {?} c
         * @return {?}
         */
        matRangeDatepickerInput.prototype.validate = function (c) {
            return this._validator ? this._validator(c) : null;
        };
        /**
         * @deprecated
         * \@deletion-target 7.0.0 Use `getConnectedOverlayOrigin` instead
         * @return {?}
         */
        matRangeDatepickerInput.prototype.getPopupConnectionElementRef = function () {
            return this.getConnectedOverlayOrigin();
        };
        /**
         * Gets the element that the datepicker popup should be connected to.
         * @return {?} The element to connect the popup to.
         */
        matRangeDatepickerInput.prototype.getConnectedOverlayOrigin = function () {
            return this._formField ? this._formField.getConnectedOverlayOrigin() : this._elementRef;
        };
        /**
         * @param {?} value
         * @return {?}
         */
        matRangeDatepickerInput.prototype.writeValue = function (value) {
            this.value = value;
        };
        /**
         * @param {?} fn
         * @return {?}
         */
        matRangeDatepickerInput.prototype.registerOnChange = function (fn) {
            this._cvaOnChange = fn;
        };
        /**
         * @param {?} fn
         * @return {?}
         */
        matRangeDatepickerInput.prototype.registerOnTouched = function (fn) {
            this._onTouched = fn;
        };
        /**
         * @param {?} isDisabled
         * @return {?}
         */
        matRangeDatepickerInput.prototype.setDisabledState = function (isDisabled) {
            this.disabled = isDisabled;
        };
        /**
         * @param {?} event
         * @return {?}
         */
        matRangeDatepickerInput.prototype._onKeydown = function (event) {
            if (event.altKey && event.keyCode === keycodes.DOWN_ARROW) {
                this._datepicker.open();
                event.preventDefault();
            }
        };
        /**
         * @param {?} value
         * @return {?}
         */
        matRangeDatepickerInput.prototype._onInput = function (value) {
            var /** @type {?} */ date = null;
            if (this._datepicker.rangeMode) {
                var /** @type {?} */ parts = value.split('-');
                if (parts.length > 1) {
                    var /** @type {?} */ position = Math.floor(parts.length / 2);
                    var /** @type {?} */ beginDateString = parts.slice(0, position).join('-');
                    var /** @type {?} */ endDateString = parts.slice(position).join('-');
                    var /** @type {?} */ beginDate = this._dateAdapter.parse(beginDateString, this._dateFormats.parse.dateInput);
                    var /** @type {?} */ endDate = this._dateAdapter.parse(endDateString, this._dateFormats.parse.dateInput);
                    this._lastValueValid = !beginDate || !endDate || this._dateAdapter.isValid(beginDate) &&
                        this._dateAdapter.isValid(endDate);
                    beginDate = this._getValidDateOrNull(beginDate);
                    endDate = this._getValidDateOrNull(endDate);
                    if (beginDate && endDate) {
                        date = /** @type {?} */ ({ begin: beginDate, end: endDate });
                    }
                }
            }
            else {
                date = this._dateAdapter.parse(value, this._dateFormats.parse.dateInput);
                this._lastValueValid = !date || this._dateAdapter.isValid(date);
                date = this._getValidDateOrNull(date);
            }
            this._value = date;
            this._cvaOnChange(date);
            this._valueChange.emit(date);
            this.dateInput.emit(new matRangeDatepickerInputEvent(this, this._elementRef.nativeElement));
        };
        /**
         * @return {?}
         */
        matRangeDatepickerInput.prototype._onChange = function () {
            this.dateChange.emit(new matRangeDatepickerInputEvent(this, this._elementRef.nativeElement));
        };
        /**
         * Returns the palette used by the input's form field, if any.
         * @return {?}
         */
        matRangeDatepickerInput.prototype._getThemePalette = function () {
            return this._formField ? this._formField.color : undefined;
        };
        /**
         * Handles blur events on the input.
         * @return {?}
         */
        matRangeDatepickerInput.prototype._onBlur = function () {
            // Reformat the input only if we have a valid value.
            if (this.value) {
                this._formatValue(this.value);
            }
            this._onTouched();
        };
        /**
         * Formats a value and sets it on the input element.
         * @param {?} value
         * @return {?}
         */
        matRangeDatepickerInput.prototype._formatValue = function (value) {
            if (value && value.hasOwnProperty('begin') && value.hasOwnProperty('end')) {
                value = /** @type {?} */ (value);
                this._elementRef.nativeElement.value =
                    value && value.begin && value.end
                        ? this._dateAdapter.format(value.begin, this._dateFormats.display.dateInput) +
                            ' - ' +
                            this._dateAdapter.format(value.end, this._dateFormats.display.dateInput)
                        : '';
            }
            else {
                value = /** @type {?} */ (value);
                this._elementRef.nativeElement.value =
                    value ? this._dateAdapter.format(value, this._dateFormats.display.dateInput) : '';
            }
        };
        /**
         * @param {?} obj The object to check.
         * @return {?} The given object if it is both a date instance and valid, otherwise null.
         */
        matRangeDatepickerInput.prototype._getValidDateOrNull = function (obj) {
            return (this._dateAdapter.isDateInstance(obj) && this._dateAdapter.isValid(obj)) ? obj : null;
        };
        return matRangeDatepickerInput;
    }());
    matRangeDatepickerInput.decorators = [
        { type: core.Directive, args: [{
                    selector: 'input[matRangeDatepicker]',
                    providers: [
                        MAT_DATEPICKER_VALUE_ACCESSOR,
                        MAT_DATEPICKER_VALIDATORS,
                        { provide: input.MAT_INPUT_VALUE_ACCESSOR, useExisting: matRangeDatepickerInput },
                    ],
                    host: {
                        '[attr.aria-haspopup]': 'true',
                        '[attr.aria-owns]': '(_datepicker?.opened && _datepicker.id) || null',
                        '[attr.min]': 'min ? _dateAdapter.toIso8601(min) : null',
                        '[attr.max]': 'max ? _dateAdapter.toIso8601(max) : null',
                        '[disabled]': 'disabled',
                        '(input)': '_onInput($event.target.value)',
                        '(change)': '_onChange()',
                        '(blur)': '_onBlur()',
                        '(keydown)': '_onKeydown($event)',
                    },
                    exportAs: 'matDatepickerInput',
                },] },
    ];
    /** @nocollapse */
    matRangeDatepickerInput.ctorParameters = function () {
        return [
            { type: core.ElementRef, },
            { type: DateAdapter, decorators: [{ type: core.Optional },] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [MAT_DATE_FORMATS,] },] },
            { type: formField.MatFormField, decorators: [{ type: core.Optional },] },
        ];
    };
    matRangeDatepickerInput.propDecorators = {
        "matRangeDatepicker": [{ type: core.Input },],
        "matDatepickerFilter": [{ type: core.Input },],
        "value": [{ type: core.Input },],
        "min": [{ type: core.Input },],
        "max": [{ type: core.Input },],
        "disabled": [{ type: core.Input },],
        "dateChange": [{ type: core.Output },],
        "dateInput": [{ type: core.Output },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * Can be used to override the icon of a `matDatepickerToggle`.
     */
    var matRangeDatepickerToggleIcon = (function () {
        function matRangeDatepickerToggleIcon() {
        }
        return matRangeDatepickerToggleIcon;
    }());
    matRangeDatepickerToggleIcon.decorators = [
        { type: core.Directive, args: [{
                    selector: '[matDatepickerToggleIcon]',
                },] },
    ];
    /**
     * @template D
     */
    var matRangeDatepickerToggle = (function () {
        /**
         * @param {?} _intl
         * @param {?} _changeDetectorRef
         */
        function matRangeDatepickerToggle(_intl, _changeDetectorRef) {
            this._intl = _intl;
            this._changeDetectorRef = _changeDetectorRef;
            this._stateChanges = rxjs.Subscription.EMPTY;
        }
        Object.defineProperty(matRangeDatepickerToggle.prototype, "disabled", {
            /**
             * Whether the toggle button is disabled.
             * @return {?}
             */
            get: function () {
                return this._disabled === undefined ? this.datepicker.disabled : !!this._disabled;
            },
            /**
             * @param {?} value
             * @return {?}
             */
            set: function (value) {
                this._disabled = coercion.coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} changes
         * @return {?}
         */
        matRangeDatepickerToggle.prototype.ngOnChanges = function (changes) {
            if (changes["datepicker"]) {
                this._watchStateChanges();
            }
        };
        /**
         * @return {?}
         */
        matRangeDatepickerToggle.prototype.ngOnDestroy = function () {
            this._stateChanges.unsubscribe();
        };
        /**
         * @return {?}
         */
        matRangeDatepickerToggle.prototype.ngAfterContentInit = function () {
            this._watchStateChanges();
        };
        /**
         * @param {?} event
         * @return {?}
         */
        matRangeDatepickerToggle.prototype._open = function (event) {
            if (this.datepicker && !this.disabled) {
                this.datepicker.open();
                event.stopPropagation();
            }
        };
        /**
         * @return {?}
         */
        matRangeDatepickerToggle.prototype._watchStateChanges = function () {
            var _this = this;
            var /** @type {?} */ datepickerDisabled = this.datepicker ? this.datepicker._disabledChange : rxjs.of();
            var /** @type {?} */ inputDisabled = this.datepicker && this.datepicker._datepickerInput ?
                this.datepicker._datepickerInput._disabledChange : rxjs.of();
            var /** @type {?} */ datepickerToggled = this.datepicker ?
                rxjs.merge(this.datepicker.openedStream, this.datepicker.closedStream) :
                rxjs.of();
            this._stateChanges.unsubscribe();
            this._stateChanges = rxjs.merge(this._intl.changes, datepickerDisabled, inputDisabled, datepickerToggled).subscribe(function () { return _this._changeDetectorRef.markForCheck(); });
        };
        return matRangeDatepickerToggle;
    }());
    matRangeDatepickerToggle.decorators = [
        { type: core.Component, args: [{
                    moduleId: module.id,
                    selector: 'mat-range-datepicker-toggle',
                    template: "<button mat-icon-button type=\"button\" [attr.aria-label]=\"_intl.openCalendarLabel\"\n        [disabled]=\"disabled\" (click)=\"_open($event)\">\n\n    <svg\n            *ngIf=\"!_customIcon\"\n            class=\"mat-datepicker-toggle-default-icon\"\n            viewBox=\"0 0 24 24\"\n            width=\"24px\"\n            height=\"24px\"\n            fill=\"currentColor\"\n            focusable=\"false\">\n        <path d=\"M0 0h24v24H0z\" fill=\"none\"/>\n        <path d=\"M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z\"/>\n    </svg>\n\n    <ng-content select=\"[matDatepickerToggleIcon]\"></ng-content>\n</button>\n",
                    styles: [".mat-form-field-appearance-legacy .mat-form-field-prefix .mat-datepicker-toggle-default-icon,.mat-form-field-appearance-legacy .mat-form-field-suffix .mat-datepicker-toggle-default-icon{width:1em}.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-prefix .mat-datepicker-toggle-default-icon,.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-suffix .mat-datepicker-toggle-default-icon{display:block;width:1.5em;height:1.5em}.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-prefix .mat-icon-button .mat-datepicker-toggle-default-icon,.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-suffix .mat-icon-button .mat-datepicker-toggle-default-icon{margin:auto}"],
                    host: {
                        'class': 'mat-datepicker-toggle',
                        '[class.mat-datepicker-toggle-active]': 'datepicker && datepicker.opened',
                    },
                    exportAs: 'matDatepickerToggle',
                    encapsulation: core.ViewEncapsulation.None,
                    preserveWhitespaces: false,
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    matRangeDatepickerToggle.ctorParameters = function () {
        return [
            { type: matRangeDatepickerIntl, },
            { type: core.ChangeDetectorRef, },
        ];
    };
    matRangeDatepickerToggle.propDecorators = {
        "datepicker": [{ type: core.Input, args: ['for',] },],
        "_customIcon": [{ type: core.ContentChild, args: [matRangeDatepickerToggleIcon,] },],
        "disabled": [{ type: core.Input },],
    };
    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MatRangeDatepickerModule = (function () {
        function MatRangeDatepickerModule() {
        }
        return MatRangeDatepickerModule;
    }());
    MatRangeDatepickerModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        button.MatButtonModule,
                        dialog.MatDialogModule,
                        overlay.OverlayModule,
                        a11y.A11yModule,
                        portal.PortalModule,
                    ],
                    exports: [
                        SatCalendar,
                        SatCalendarBody,
                        matRangeDatepicker,
                        matRangeDatepickerContent,
                        matRangeDatepickerInput,
                        matRangeDatepickerToggle,
                        matRangeDatepickerToggleIcon,
                        SatMonthView,
                        SatYearView,
                        SatMultiYearView,
                        SatCalendarHeader,
                        RangeCalendar,
                    ],
                    declarations: [
                        SatCalendar,
                        SatCalendarBody,
                        RangeCalendar,
                        matRangeDatepicker,
                        matRangeDatepickerContent,
                        matRangeDatepickerInput,
                        matRangeDatepickerToggle,
                        matRangeDatepickerToggleIcon,
                        SatMonthView,
                        SatYearView,
                        SatMultiYearView,
                        SatCalendarHeader,
                    ],
                    providers: [
                        matRangeDatepickerIntl,
                    ],
                    entryComponents: [
                        matRangeDatepickerContent,
                        SatCalendarHeader,
                    ]
                },] },
    ];

    exports.NativeDateModule = NativeDateModule;
    exports.MatRangeNativeDateModule = MatRangeNativeDateModule;
    exports.MAT_DATE_LOCALE = MAT_DATE_LOCALE;
    exports.MAT_DATE_LOCALE_FACTORY = MAT_DATE_LOCALE_FACTORY;
    exports.MAT_DATE_LOCALE_PROVIDER = MAT_DATE_LOCALE_PROVIDER;
    exports.DateAdapter = DateAdapter;
    exports.MAT_DATE_FORMATS = MAT_DATE_FORMATS;
    exports.NativeDateAdapter = NativeDateAdapter;
    exports.MAT_NATIVE_DATE_FORMATS = MAT_NATIVE_DATE_FORMATS;
    exports.MatRangeDatepickerModule = MatRangeDatepickerModule;
    exports.SatCalendar = SatCalendar;
    exports.SatCalendarHeader = SatCalendarHeader;
    exports.SatCalendarCell = SatCalendarCell;
    exports.SatCalendarBody = SatCalendarBody;
    exports.MAT_DATEPICKER_SCROLL_STRATEGY = MAT_DATEPICKER_SCROLL_STRATEGY;
    exports.MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY = MAT_DATEPICKER_SCROLL_STRATEGY_FACTORY;
    exports.matRangeDatepickerContentBase = matRangeDatepickerContentBase;
    exports._matRangeDatepickerContentMixinBase = _matRangeDatepickerContentMixinBase;
    exports.matRangeDatepickerContent = matRangeDatepickerContent;
    exports.matRangeDatepicker = matRangeDatepicker;
    exports.matDatepickerAnimations = matDatepickerAnimations;
    exports.MAT_DATEPICKER_VALUE_ACCESSOR = MAT_DATEPICKER_VALUE_ACCESSOR;
    exports.MAT_DATEPICKER_VALIDATORS = MAT_DATEPICKER_VALIDATORS;
    exports.matRangeDatepickerInputEvent = matRangeDatepickerInputEvent;
    exports.matRangeDatepickerInput = matRangeDatepickerInput;
    exports.matRangeDatepickerIntl = matRangeDatepickerIntl;
    exports.matRangeDatepickerToggleIcon = matRangeDatepickerToggleIcon;
    exports.matRangeDatepickerToggle = matRangeDatepickerToggle;
    exports.SatMonthView = SatMonthView;
    exports.SatYearView = SatYearView;
    exports.ɵa = SatMultiYearView;
    exports.ɵb = RangeCalendar;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
